

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Module 1, Practical 10 &mdash; QCB Scientific Programming Lab</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script type="text/javascript" src="_static/js/jupman.js"></script>
        <script type="text/javascript" src="_static/js/pytutor-embed.bundle.min.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/jupman.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/jupman-web.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Module 1, Practical 10b" href="M1_practical10b.html" />
    <link rel="prev" title="Module 1, Practical 9" href="M1_practical9.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="toc-page.html" class="icon icon-home"> QCB Scientific Programming Lab
          

          
          </a>

          
            
            
              <div class="version">
                dev
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Scientific Programming for QCB</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#General-Info">General Info</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#Timetable-and-lecture-rooms">Timetable and lecture rooms</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Zoom-links">Zoom links</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Equipment">Equipment</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical1.html">Module 1, Practical 1</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Setting-up-the-environment">Setting up the environment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical1.html#Linux-on-windows">Linux on windows</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical1.html#A-dual-boot-system">A dual boot system</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Our-toolbox">Our toolbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Installing-Python3-in-Linux">Installing Python3 in Linux</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Installing-Python3-in-Windows/Mac">Installing Python3 in Windows/Mac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical1.html#OPTION-1:">OPTION 1:</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical1.html#OPTION-2-(easier):">OPTION 2 (easier):</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#The-console">The console</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Visual-Studio-Code">Visual Studio Code</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#The-debugger">The debugger</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#A-quick-Jupyter-primer-(just-for-your-information,-skip-if-not-interested)">A quick Jupyter primer (just for your information, skip if not interested)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical1.html#Installation">Installation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical1.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical2.html">Module 1, Practical 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Modules">Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Objects">Objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Variables">Variables</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Numeric-types">Numeric types</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical2.html#Integers">Integers</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical2.html#Booleans">Booleans</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical2.html#Real-numbers">Real numbers</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Strings">Strings</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical2.html#Methods-for-the-str-object">Methods for the str object</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical2.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical3.html">Module 1, Practical 3</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical3.html#Lists">Lists</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical3.html#Operators-for-lists">Operators for lists</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical3.html#Methods-of-the-class-list">Methods of the class list</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical3.html#From-strings-to-lists,-the-split-method">From strings to lists, the <code class="docutils literal notranslate"><span class="pre">split</span></code> method</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical3.html#And-back-to-strings-with-the-join-method">And back to strings with the <code class="docutils literal notranslate"><span class="pre">join</span></code> method</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical3.html#Tuples">Tuples</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical3.html#Introduction-to-sets">Introduction to sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical3.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical4.html">Module 1, Practical 4</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical4.html#Execution-flow">Execution flow</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical4.html#Conditionals">Conditionals</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#The-basic-if---else-statement">The basic <em>if - else</em> statement</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#The-if---elif---else-statement">The <em>if - elif - else</em> statement</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical4.html#Loops">Loops</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#For-loop">For loop</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#Looping-over-a-range">Looping over a range</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#Nested-for-loops">Nested for loops</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical4.html#While-loops">While loops</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical4.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical5.html">Module 1, Practical 5</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical5.html#More-on-loops">More on loops</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical5.html#Ternary-operator">Ternary operator</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical5.html#Break-and-continue">Break and continue</a><ul>
<li class="toctree-l4"><a class="reference internal" href="M1_practical5.html#Continue-statement">Continue statement</a></li>
<li class="toctree-l4"><a class="reference internal" href="M1_practical5.html#Break-statement">Break statement</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical5.html#List-comprehension">List comprehension</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical5.html#Dictionaries">Dictionaries</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical5.html#Functions-working-on-dictionaries">Functions working on dictionaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical5.html#Dictionary-methods">Dictionary methods</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical5.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical6.html">Module 1, Practical 6</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical6.html#Functions">Functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical6.html#Namespace-and-variable-scope">Namespace and variable scope</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical6.html#Argument-passing">Argument passing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Positional-arguments">Positional arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Passing-arguments-by-keyword">Passing arguments by keyword</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Specifying-default-values">Specifying default values</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical6.html#File-input-and-output">File input and output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Opening-a-file">Opening a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Closing-a-file">Closing a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Reading-from-a-file">Reading from a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#Writing-to-a-file">Writing to a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical6.html#String-formatting-with-format">String formatting with format</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical6.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical7.html">Module 1, Practical 7</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical7.html#Gentle-reminder-on-functions">Gentle reminder on functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical7.html#Getting-input-from-the-command-line">Getting input from the command line</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical7.html#Argparse">Argparse</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical7.html#Libraries-installation-for-the-next-practicals">Libraries installation for the next practicals</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical7.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical8.html">Module 1, Practical 8</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Libraries-installation">Libraries installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Pandas">Pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Series">Series</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#How-to-define-and-access-a-Series">How to define and access a Series</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Operator-broadcasting">Operator broadcasting</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Filtering">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Missing-data">Missing data</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Computing-stats">Computing stats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Plotting-data">Plotting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Pandas-DataFrames">Pandas DataFrames</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Define-a-DataFrame">Define a DataFrame</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Loading-data-from-external-files">Loading data from external files</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Extract-values-by-row-and-column">Extract values by row and column</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Broadcasting,-filtering-and-computing-stats">Broadcasting, filtering and computing stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Merging-DataFrames">Merging DataFrames</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical8.html#Grouping">Grouping</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical8.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical9.html">Module 1, Practical 9</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Numpy">Numpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Numpy-ndarray">Numpy ndarray</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Random-arrays">Random arrays</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Numpy-to-and-from-pandas">Numpy to and from pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Reshaping">Reshaping</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Iterating-over-arrays-and-Indexing">Iterating over arrays and Indexing</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Broadcasting-and-arithmetic-functions">Broadcasting and arithmetic functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Filtering">Filtering</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Linear-algebra">Linear algebra</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Matplotlib">Matplotlib</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical9.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Module 1, Practical 10</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Biopython">Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Installation-of-Biopython">Installation of Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="#General-overview">General overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Sequence-objects">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Looping,-slicing-and-concatenation">Looping, slicing and concatenation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Complement-and-reverse-complement">Complement and reverse complement</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Transcription-and-translation">Transcription and translation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#MutableSeq">MutableSeq</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Sequence-annotation">Sequence annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Sequence-I/O">Sequence I/O</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#BioSeqIO.parse()">BioSeqIO.parse()</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Sequences-as-dictionaries">Sequences as dictionaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Writing-sequence-files">Writing sequence files</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Multiple-sequence-alignment">Multiple sequence alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Parsing-multiple-sequence-alignments">Parsing multiple sequence alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Writing-multiple-sequence-alignments">Writing multiple sequence alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Manipulating-alignments">Manipulating alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Biopython’s-pairwise2-alignment">Biopython’s pairwise2 alignment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical10b.html">Module 1, Practical 10b</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#Biopython">Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#BLAST">BLAST</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M1_practical10b.html#The-function-qblast">The function qblast</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical10b.html#Parsing-qblast-output">Parsing qblast output</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical10b.html#Saving-results-to-file">Saving results to file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical10b.html#Open-a-blast-.XML-file">Open a blast .XML file</a></li>
<li class="toctree-l3"><a class="reference internal" href="M1_practical10b.html#The-BLAST-record-class">The BLAST record class</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#Getting-data-from-NCBI">Getting data from NCBI</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#Getting-data-from-ExPASy">Getting data from ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#3D-structure-and-PDB">3D structure and PDB</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical10b.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_practical11.html">Module 1, Practical 11</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_practical11.html#Recursive-functions">Recursive functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_practical11.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_midTermSimulation.html">Module 1, MidTerm Simulation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_midTermSimulation.html#Available-material-for-exams">Available material for exams</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTermSimulation.html#Download-the-data">Download the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTermSimulation.html#Problem-1">Problem 1</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTermSimulation.html#Problem-2">Problem 2</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M1_midTerm.html">Module 1, MidTerm</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#Download-the-data">Download the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#Problem-1">Problem 1</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#Problem-2">Problem 2</a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#2.1)-computeKmers(seq,-kmerLen,-kmerDict)">2.1) <code class="docutils literal notranslate"><span class="pre">computeKmers(seq,</span> <span class="pre">kmerLen,</span> <span class="pre">kmerDict)</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#2.2)-parseFasta(fileName,kmerLen)">2.2) <code class="docutils literal notranslate"><span class="pre">parseFasta(fileName,kmerLen)</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="M1_midTerm.html#2.3)-plotKmerSpectrum(kmerDict)">2.3) <code class="docutils literal notranslate"><span class="pre">plotKmerSpectrum(kmerDict)</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical1.html">Module 2, Practical 1</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Object-Oriented-Programming">Object Oriented Programming</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical1.html#Classes,-Methods-and-Objects">Classes, Methods and Objects</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical1.html#Self">Self</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical1.html#Definition-of-a-class">Definition of a class</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical1.html#The-life-cycle-of-classes-and-objects-in-a-program:">The life-cycle of classes and objects in a program:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Encapsulation">Encapsulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Special-methods">Special methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Inheritance-and-overriding">Inheritance and overriding</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical1.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Lambda-functions">Lambda functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical1.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical2.html">Module 2, Practical 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical2.html#Testing">Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical2.html#Unit-testing">Unit testing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical2.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical2.html#Measuring-time">Measuring time</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical2.html#id1">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical2.html#Measuring-memory">Measuring memory</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical2.html#id2">Exercise</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical3.html">Module 2, Practical 3</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical3.html#Complexity">Complexity</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical3.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical4.html">Module 2, Practical 4</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical4.html#Regular-expressions">Regular expressions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#Basic-Patterns">Basic Patterns</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#Group-extraction">Group extraction</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#Exercises">Exercises</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#re.compile">re.compile</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#Substitution">Substitution</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical4.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical4.html#id1">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical5.html">Module 2, Practical 5</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Sorting-algorithms">Sorting algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Selection-sort">Selection sort</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#A-base-class-for-sorting-algorithms">A base class for sorting algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#Exercise-(implementation)">Exercise (implementation)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Insertion-sort">Insertion sort</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#id1">Exercise (implementation)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Merge-sort-and-Quick-sort">Merge sort and Quick sort</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#Merge-sort">Merge sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#Exercise-(Implementation)">Exercise (Implementation)</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#Quick-sort">Quick sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#id3">Exercise (Implementation)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Exercise-(algorithm-benchmark)">Exercise (algorithm benchmark)</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Reflection">Reflection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical5.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical5.html#Exercise-(Counting-Sort)">Exercise (Counting Sort)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical6.html">Module 2, Practical 6</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical6.html#Data-structures">Data structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Abstract-Data-Types-(ADTs)">Abstract Data Types (ADTs)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="M2_practical6.html#ADT:-Sequence">ADT: Sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="M2_practical6.html#ADT:-Set">ADT: Set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Exercise-(set-implementation)">Exercise (set implementation)</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#ADT:-dictionary">ADT: dictionary</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical6.html#Linked-lists">Linked lists</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Example:-bidirectional-linked-list">Example: bidirectional linked list</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Exercise-(complete-bidirectional-list)">Exercise (complete bidirectional list)</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Exercise-(circular-list)">Exercise (circular list)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical6.html#Stacks">Stacks</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical6.html#Queues">Queues</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical6.html#Collections">Collections</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical6.html#id1">Exercise</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical7.html">Module 2, Practical 7</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical7.html#Trees">Trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical7.html#ADT:-Binary-Tree">ADT: Binary Tree</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical7.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical7.html#Tree-visit">Tree visit</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical7.html#Binary-Tree-visits:-DFS">Binary Tree visits: DFS</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical7.html#Binary-Tree-visits:-BFS">Binary Tree visits: BFS</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical7.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical7.html#ADT:-Generic-Tree">ADT: Generic Tree</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical7.html#Exercise-(implementation)">Exercise (implementation)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical8.html">Module 2, Practical 8</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical8.html#Graphs">Graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical8.html#Graph-ADT">Graph ADT</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical8.html#Implementation-as-adjacency-matrix">Implementation as adjacency matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical8.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical8.html#Implementation-as-(adjacency)-linked-list">Implementation as (adjacency) linked list</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical8.html#id1">Exercises</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical9.html">Module 2, Practical 9</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical9.html#Graphs-recap">Graphs recap</a></li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical9.html#Visits">Visits</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical9.html#Depth-First-Search-(DFS)">Depth First Search (DFS)</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical9.html#Exercises">Exercises</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical9.html#Breadth-First-Search-(BFS)">Breadth First Search (BFS)</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_practical9.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical9.html#id1">Exercise</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_practical10.html">Module 2, Practical 10</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_practical10.html#Dynamic-programming">Dynamic programming</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical10.html#Excercise">Excercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical10.html#Finding-(the-optimal)-overlaps-among-DNA-strings">Finding (the optimal) overlaps among DNA strings</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical10.html#Exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_practical10.html#Greedy-programming">Greedy programming</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_practical10.html#Exercises">Exercises</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="M2_midTermSimulation.html">Module 2, MidTerm Simulation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="M2_midTermSimulation.html#Theoretical-part">Theoretical part</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_midTermSimulation.html#Exercise-1">Exercise 1</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_midTermSimulation.html#Exercise-2">Exercise 2</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="M2_midTermSimulation.html#Practical-part">Practical part</a><ul>
<li class="toctree-l3"><a class="reference internal" href="M2_midTermSimulation.html#Exercise-3">Exercise 3</a></li>
<li class="toctree-l3"><a class="reference internal" href="M2_midTermSimulation.html#Exercise-4">Exercise 4</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="toc.html"> </a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="toc-page.html">QCB Scientific Programming Lab</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          


    <div id='jupman-langs'></div>

    <!-- SMALL TEMPLATE FOR LANGUAGES (VERY CUSTOM FOR softpython.org !!)
    <script>
    var hn = window.location.hostname;
    var pn = window.location.pathname;

    if (hn.includes('softpython.org')){
        JUPMAN_LANG = hn.slice(0,window.location.hostname.indexOf('.'));
    } else if (pn.includes('/_build/html')){
        var pos = pn.indexOf('/_build/html');
        JUPMAN_LANG = pn.slice(pos-2, pos);
    } else {
        JUPMAN_LANG = 'en';
    }
    console.log('JUPMAN_LANG=',JUPMAN_LANG);
    </script>
    -->
    
    
     <!-- <li> <h2 style="background-color:red; color:yellow">WARNING! THIS VERSION IS OLD !  <br/>
    <a style=" text-decoration: underline;" href="https://jupman.readthedocs.io/en/latest">GOT TO LATEST VERSION</a></li>
 </h2></li>-->
    <!-- Commented because doesn't work with sections,
         see https://github.com/DavidLeoni/jupman/issues/39
    <li><a href="toc-page.html">Docs</a> &raquo;</li>
    
        <li><a href="toc.html"> </a> &raquo;</li>
    
    <li>Module 1, Practical 10</li>
    -->

          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Module-1,-Practical-10">
<h1>Module 1, Practical 10<a class="headerlink" href="#Module-1,-Practical-10" title="Permalink to this headline">¶</a></h1>
<p>In this practical we will get acquainted with Biopython.</p>
<div class="section" id="Biopython">
<h2>Biopython<a class="headerlink" href="#Biopython" title="Permalink to this headline">¶</a></h2>
<p>From the Biopython tutorial: The Biopython Project is an international association of developers of freely available Python tools for computational molecular biology.</p>
<p>The goal of Biopython is <strong>to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes.</strong> Biopython features include parsers for various Bioinformatics file formats (FASTA, FASTQ, BLAST, PDB, Clustalw, Genbank,…), access to online services (NCBI, Expasy,…), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS…), a standard sequence class, various clustering modules, a KD tree data structure etc. and even some
documentation :-).</p>
<p>In this practical we will see some features of Biopython but please refer to <a class="reference external" href="http://biopython.org/wiki/Documentation">biopython documentation</a> to discover all its features, recipes etc.</p>
<p>These notes are largely based on the tutorial that is available <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">here</a>.</p>
</div>
<div class="section" id="Installation-of-Biopython">
<h2>Installation of Biopython<a class="headerlink" href="#Installation-of-Biopython" title="Permalink to this headline">¶</a></h2>
<p>To test if Biopython is installed in your system you can import the library with:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>import Bio
</pre></div>
</div>
<p>if the execution gives:</p>
<img alt="imp1" src="_images/import.png" />
<p>Biopython needs to be installed.</p>
<p><strong>In windows</strong> installing Biopython should be as easy as opening the command prompt as admininstrator (typing <code class="docutils literal notranslate"><span class="pre">cmd</span></code> and then right clicking on the link choosing run as admininstrator) and then <code class="docutils literal notranslate"><span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">biopython</span></code>. If that does not work, you should be able to install biopython with the following: <code class="docutils literal notranslate"><span class="pre">python3.xx</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code> (where you should change python3.xx with the path to your python3.xx executable).</p>
<p><strong>In linux</strong> <code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">biopython</span></code> will install biopython for python3 up to python3.5. On python 3.6 and later the command is: <code class="docutils literal notranslate"><span class="pre">python3.6</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code> or, more in general, <code class="docutils literal notranslate"><span class="pre">pythonX.X</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code>.</p>
<p>Detailed installation instructions can be found <a class="reference external" href="https://github.com/biopython/biopython">here</a>.</p>
</div>
<div class="section" id="General-overview">
<h2>General overview<a class="headerlink" href="#General-overview" title="Permalink to this headline">¶</a></h2>
<p>Biopython provides the capability of parsing standard bioinformatics formats into python usable data structures. Some of the supported tools and formats are:</p>
<ol class="arabic simple">
<li><p>Blast output – both from standalone and WWW Blast</p></li>
<li><p>Clustalw</p></li>
<li><p>FASTA</p></li>
<li><p>GenBank</p></li>
<li><p>PubMed and Medline</p></li>
<li><p>ExPASy files, like Enzyme and Prosite</p></li>
<li><p>SCOP, including ‘dom’ and ‘lin’ files</p></li>
<li><p>UniGene</p></li>
<li><p>SwissProt</p></li>
</ol>
<p>among the others. In particular it provides a sequence class to perform the most common operations on sequences, parsing of alignments, clustering etc. and it can quickly connect to the most popular databases to download data or parse information on the fly.</p>
</div>
<div class="section" id="Sequence-objects">
<h2>Sequence objects<a class="headerlink" href="#Sequence-objects" title="Permalink to this headline">¶</a></h2>
<p>Biopython has a specific object <code class="docutils literal notranslate"><span class="pre">Seq</span></code> to deal with biological sequences which is a lot more powerful than the <code class="docutils literal notranslate"><span class="pre">str</span></code> object of python when it comes to manage biological data. The object <strong>Seq</strong> has methods like <strong>translate()</strong> and <strong>reverse_complement()</strong> which are very handy.</p>
<p>As in the case of python’s <code class="docutils literal notranslate"><span class="pre">str</span></code> objects, the object <code class="docutils literal notranslate"><span class="pre">Seq</span></code> is immutable. The mutable version is <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code>.</p>
<p>All the information on Seq objects can be found <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html">here</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">s</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTACATAATA&quot;</span><span class="p">)</span>
<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTAC&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;S:&quot;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;dna_seq:&quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">)</span>


<span class="n">my_prot</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MGNAAAAKKGSEQE&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;my_prot:&quot;</span><span class="p">,</span> <span class="n">my_prot</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
S: GATTACATAATA
dna_seq: GATTATACGTAC
my_prot: MGNAAAAKKGSEQE
</pre></div></div>
</div>
</div>
<div class="section" id="Looping,-slicing-and-concatenation">
<h2>Looping, slicing and concatenation<a class="headerlink" href="#Looping,-slicing-and-concatenation" title="Permalink to this headline">¶</a></h2>
<p>Note also that <strong>Seq objects behave like strings</strong>, and the consistency of the alphabet (i.e. if the sequence is DNA, RNA, aminoacid sequence…) is not checked anymore in the latest version. We can concatenate a sequence of DNA with that of aminoacids…</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTAC&quot;</span><span class="p">)</span>
<span class="n">my_prot</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MGNAAAAKKGSEQE&quot;</span><span class="p">)</span>


<span class="c1">#Does it really make sense though?!?</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dna_seq</span> <span class="o">+</span> <span class="n">my_prot</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GATTATACGTACMGNAAAAKKGSEQE
</pre></div></div>
</div>
<p>We can also loop through the elements and slice strings with the usual [S:E:step] operator (and as usual S is included, E is excluded). It is also possible to convert a Seq object into a string.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTACGGCTA&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">dna_seq</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">base</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">sub_seq</span> <span class="o">=</span> <span class="n">dna_seq</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">)</span>

<span class="c1">#Let&#39;s reverse the string:</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reversed: &quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="c1">#from Seq to string:</span>
<span class="n">dna_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;As string:&quot;</span><span class="p">,</span> <span class="n">dna_str</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">dna_str</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
G A T T A T A C G T A C G G C T A
ATACGT
Reversed:  ATCGGCATGCATATTAG
As string: GATTATACGTACGGCTA
&lt;class &#39;str&#39;&gt;
</pre></div></div>
</div>
<p>The object Seq provides several methods similar to the corresponding methods of strings:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.count(s)</span></code> : counts the number of times s appears in the sequence;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.upper()</span></code> : makes the sequence of the object Seq upper case</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.lower()</span></code> : makes the sequence of the object Seq lower case</p></li>
</ol>
<p><strong>Example:</strong> Let’s compute the GC content of the sequence “GATTRWWACGTACGGCTASATTACSCCGGCTA”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">GC</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTRWWACGTACGGCTASATTACSCCGGCTA&quot;</span><span class="p">)</span>

<span class="n">gc</span> <span class="o">=</span> <span class="p">(</span><span class="n">dna_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">dna_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">))</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GC </span><span class="si">% i</span><span class="s2">s </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">gc</span><span class="p">))</span>

<span class="n">correct_gc</span> <span class="o">=</span> <span class="n">GC</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GC </span><span class="si">% i</span><span class="s2">s </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">correct_gc</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;original:&quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;lower_case:&quot;</span><span class="p">,</span><span class="n">dna_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GC % is 40.62
GC % is 46.88
original: GATTRWWACGTACGGCTASATTACSCCGGCTA
lower_case: gattrwwacgtacggctasattacsccggcta
</pre></div></div>
</div>
<p>The difference in the results goes down to the fact that Biopython’s GC method takes into account ambiguous nucleotides like (S which stands for C or G <a class="reference external" href="https://www.bioinformatics.org/sms/iupac.html">ref iupac</a>). This is a method provided in the <strong>SeqUtils</strong> module (more info on all the methods <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqUtils-module.html">here</a>).</p>
<div class="section" id="Complement-and-reverse-complement">
<h3>Complement and reverse complement<a class="headerlink" href="#Complement-and-reverse-complement" title="Permalink to this headline">¶</a></h3>
<p>A usual operation on DNA strings is the reverse complement. Biopython provides two specific methods to do that:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.complement()</span></code> to complement the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.reverse_complement()</span></code> to reverse complement the sequence.</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Original sequence:</span><span class="se">\t</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span> <span class="p">)</span>
<span class="n">comp</span>  <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Complement:</span><span class="se">\t\t</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">comp</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">revcomp</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reverse complement:</span><span class="se">\t</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">revcomp</span><span class="p">))</span>


</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Original sequence:      GATCGATGGGCCTATATAGGATCGAAAATCGC

Complement:             CTAGCTACCCGGATATATCCTAGCTTTTAGCG

Reverse complement:     GCGATTTTCGATCCTATATAGGCCCATCGATC
</pre></div></div>
</div>
</div>
<div class="section" id="Transcription-and-translation">
<h3>Transcription and translation<a class="headerlink" href="#Transcription-and-translation" title="Permalink to this headline">¶</a></h3>
<p>Remember that:</p>
<img alt="tt" src="_images/trans_transl.png" />
<p>[taken from the Biopython tutorial].</p>
<p>Biopython has a method to transcribe a DNA sequence into a mRNA sequence (and one to go back):</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.transcribe()</span></code> that transcribes a DNA sequence into mRNA (converting Ts in Us).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.back_transcribe()</span></code> that back-transcribes a mRNA sequence into DNA (converting Us in Ts).</p></li>
</ol>
<p>Note that transcribing a protein will result in a <strong>TranslationError</strong>.</p>
<p>Alike, a DNA or mRNA sequence can be translated into a protein with:</p>
<p><code class="docutils literal notranslate"><span class="pre">Seq.translate(table=N,</span> <span class="pre">to_stop=True/False,</span> <span class="pre">stop_symbol</span> <span class="pre">=symbol)</span></code> that converts a DNA sequence or mRNA into the corresponding protein.</p>
<p>Optionally, the translate method accepts a parameter <code class="docutils literal notranslate"><span class="pre">table</span></code> to specify the NCBI translation table used (see below for codes). <code class="docutils literal notranslate"><span class="pre">to_stop</span></code> makes the translation stop at the first stop codon if True (<strong>in this case the stop codon is not translated!</strong>), <code class="docutils literal notranslate"><span class="pre">stop_symbol</span></code> to replace the asterisk with any other stop character (Es. “&#64;”).</p>
<p><img alt="ncbitab" src="_images/ncbi_table.png" /> Stop codons will be represented by an asterisk.</p>
<p><strong>Example:</strong> Let’s transcribe into mRNA and translate into protein the following coding sequence: ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>

<span class="n">mrna</span> <span class="o">=</span> <span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;... and back&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Translation to protein:&quot;</span><span class="p">)</span>
<span class="n">prot</span> <span class="o">=</span> <span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">prot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Up to first stop:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mitocondrial translation: (TGA is W!)&quot;</span><span class="p">)</span>
<span class="n">mit_prot</span> <span class="o">=</span> <span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mit_prot</span><span class="p">)</span>
<span class="c1">#The following produces a translation error!</span>
<span class="c1">#print(&quot;RE-Translated protein: {}&quot;.format(prot.translate()))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG
AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG

... and back
ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG

Translation to protein:
MAIVMGR*KGAR*

Up to first stop:
MAIVMGR

Mitocondrial translation: (TGA is W!)
MAIVMGRWKGAR*
</pre></div></div>
</div>
<p>Translation tables can be accessed at runtime through:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Data</span> <span class="kn">import</span> <span class="n">CodonTable</span>
<span class="c1">#equivalent to:</span>
<span class="c1">#standard_table = CodonTable.unambiguous_dna_by_id[1]</span>
<span class="n">standard_table</span> <span class="o">=</span> <span class="n">CodonTable</span><span class="o">.</span><span class="n">unambiguous_dna_by_name</span><span class="p">[</span><span class="s2">&quot;Standard&quot;</span><span class="p">]</span>
<span class="c1">#equivalent to:</span>
<span class="c1">#standard_table = CodonTable.unambiguous_dna_by_id[2]</span>
<span class="n">mito_table</span> <span class="o">=</span> <span class="n">CodonTable</span><span class="o">.</span><span class="n">unambiguous_dna_by_name</span><span class="p">[</span><span class="s2">&quot;Vertebrate Mitochondrial&quot;</span><span class="p">]</span>


<span class="nb">print</span><span class="p">(</span><span class="n">standard_table</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mito_table</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Table 1 Standard, SGC0

  |  T      |  C      |  A      |  G      |
--+---------+---------+---------+---------+--
T | TTT F   | TCT S   | TAT Y   | TGT C   | T
T | TTC F   | TCC S   | TAC Y   | TGC C   | C
T | TTA L   | TCA S   | TAA Stop| TGA Stop| A
T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G
--+---------+---------+---------+---------+--
C | CTT L   | CCT P   | CAT H   | CGT R   | T
C | CTC L   | CCC P   | CAC H   | CGC R   | C
C | CTA L   | CCA P   | CAA Q   | CGA R   | A
C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G
--+---------+---------+---------+---------+--
A | ATT I   | ACT T   | AAT N   | AGT S   | T
A | ATC I   | ACC T   | AAC N   | AGC S   | C
A | ATA I   | ACA T   | AAA K   | AGA R   | A
A | ATG M(s)| ACG T   | AAG K   | AGG R   | G
--+---------+---------+---------+---------+--
G | GTT V   | GCT A   | GAT D   | GGT G   | T
G | GTC V   | GCC A   | GAC D   | GGC G   | C
G | GTA V   | GCA A   | GAA E   | GGA G   | A
G | GTG V   | GCG A   | GAG E   | GGG G   | G
--+---------+---------+---------+---------+--

Table 2 Vertebrate Mitochondrial, SGC1

  |  T      |  C      |  A      |  G      |
--+---------+---------+---------+---------+--
T | TTT F   | TCT S   | TAT Y   | TGT C   | T
T | TTC F   | TCC S   | TAC Y   | TGC C   | C
T | TTA L   | TCA S   | TAA Stop| TGA W   | A
T | TTG L   | TCG S   | TAG Stop| TGG W   | G
--+---------+---------+---------+---------+--
C | CTT L   | CCT P   | CAT H   | CGT R   | T
C | CTC L   | CCC P   | CAC H   | CGC R   | C
C | CTA L   | CCA P   | CAA Q   | CGA R   | A
C | CTG L   | CCG P   | CAG Q   | CGG R   | G
--+---------+---------+---------+---------+--
A | ATT I(s)| ACT T   | AAT N   | AGT S   | T
A | ATC I(s)| ACC T   | AAC N   | AGC S   | C
A | ATA M(s)| ACA T   | AAA K   | AGA Stop| A
A | ATG M(s)| ACG T   | AAG K   | AGG Stop| G
--+---------+---------+---------+---------+--
G | GTT V   | GCT A   | GAT D   | GGT G   | T
G | GTC V   | GCC A   | GAC D   | GGC G   | C
G | GTA V   | GCA A   | GAA E   | GGA G   | A
G | GTG V(s)| GCG A   | GAG E   | GGG G   | G
--+---------+---------+---------+---------+--
</pre></div></div>
</div>
</div>
<div class="section" id="MutableSeq">
<h3>MutableSeq<a class="headerlink" href="#MutableSeq" title="Permalink to this headline">¶</a></h3>
<p>Seq objects are <strong>immutable and therefore we cannot change their value</strong>. If we need to do so, we need to convert them into <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code> objects to avoid a runtime <strong>TypeError</strong> that will occur when we try to change a Seq. Seq to MutableSeq conversion can be done by:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.tomutable()</span></code> to convert a Seq to a MutableSeq.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">MutableSeq.toseq()</span></code> to convert a MutableSeq to a Seq.</p></li>
</ol>
<p>This will allow us to assign the value of characters within the MutableSeq with <code class="docutils literal notranslate"><span class="pre">MutableSeq[ind]</span> <span class="pre">=</span> <span class="pre">newValue</span></code>. Note also that methods like complement change the actual object if applied to a MutableSeq:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="c1">#cannot do the following because Seq is immutable</span>
<span class="c1">#coding_dna[10:20] = &quot;TATATATATA&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Coding DNA:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>
<span class="n">mut_dna</span> <span class="o">=</span> <span class="n">coding_dna</span><span class="o">.</span><span class="n">tomutable</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mutable DNA:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
<span class="n">mut_dna</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;TATATATATA&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mutable DNA now:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rev comp (Seq not changed!):&quot;</span><span class="p">)</span>
<span class="n">coding_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rev comp (MutableSeq changed!):&quot;</span><span class="p">)</span>
<span class="n">mut_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Coding DNA:
ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG
Mutable DNA:
ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG
Mutable DNA now:
ATGGCCATTGTATATATATACTGAAAGGGTGCCCGATAG

Rev comp (Seq not changed!):
ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG
Rev comp (MutableSeq changed!):
CTATCGGGCACCCTTTCAGTATATATATACAATGGCCAT
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Sequence-annotation">
<h2>Sequence annotation<a class="headerlink" href="#Sequence-annotation" title="Permalink to this headline">¶</a></h2>
<p>Sequence records are annotations associated to sequences. In biopython there is a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object to store this information. <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> are part of the <code class="docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> module which therefore needs to be imported before being used.</p>
<p>The <strong>SeqRecord</strong> class has the following attributes:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.seq</span></code> : the sequence (the Seq object)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.id</span></code> : the identifier of the sequence, typically an accession number</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.name</span></code> : a “common” name or identifier sometimes identical to the accession number</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.description</span></code> : a human readable description of the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.letter_annotations</span></code> : a per letter annotation using a restricted dictionary (e.g. quality)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.annotations</span></code> : a dictionary of unstructured annotation (e.g. organism, publications,…)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.features</span></code> : a list of SeqFeature objects with more structured information (e.g. genes pos).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.dbxrefs</span></code> : a list of database cross references.</p></li>
</ol>
<p>Information on SeqFeature objects can be found <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec38">here</a>.</p>
<p>Although you can manually create a SeqRecord, normally you will access SeqRecords populated by methods like SeqIO, <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec33">here</a> you can find how to manually create a SeqRecord.</p>
</div>
<div class="section" id="Sequence-I/O">
<h2>Sequence I/O<a class="headerlink" href="#Sequence-I/O" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> module aims to provide a simple way to work with several different sequence file formats. Detailed information on the module can be found <a class="reference external" href="http://biopython.org/wiki/SeqIO">here</a>.</p>
<p><strong>Example:</strong></p>
<p>Read a fasta file <a class="reference external" href="file_samples/NC_005816.fna">NC005816.fna</a> containing the whole sequence for Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1 and retrieve some information about the sequence.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/NC_005816.fna&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence [first 30 bases]:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">30</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The id:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The description:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The record is a: &quot;</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ID: gi|45478711|ref|NC_005816.1|
Name: gi|45478711|ref|NC_005816.1|
Description: gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence
Number of features: 0
Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;)

Sequence [first 30 bases]:
TGTAACGAACGGTGCAATAGTGATCCACAC

The id:
gi|45478711|ref|NC_005816.1|

The description:
gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence

The record is a:  &lt;class &#39;Bio.SeqRecord.SeqRecord&#39;&gt;
</pre></div></div>
</div>
<div class="admonition warning">
<p><strong>WARNING:</strong> When dealing with very large or FASTQ files, the overhead of working with all these objects can make scripts too slow. In this case <code class="docutils literal notranslate"><span class="pre">SimpleFastaParser</span></code> and <code class="docutils literal notranslate"><span class="pre">FastqGeneralIterator</span></code> parsers might be better as they return just a tuple of strings for each record.</p>
</div>
<p>For example, to use <code class="docutils literal notranslate"><span class="pre">SimpleFastaParser</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.SeqIO.FastaIO</span> <span class="kn">import</span> <span class="n">SimpleFastaParser</span>


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;file_samples/NC_005816.fna&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_handle</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SimpleFastaParser</span><span class="p">(</span><span class="n">in_handle</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="n">seq</span><span class="p">[</span><span class="mi">51</span><span class="p">:</span><span class="mi">101</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">...&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Header:</span><span class="se">\n</span><span class="si">{}</span><span class="se">\n</span><span class="s2">Sequence:</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">title</span><span class="p">,</span><span class="n">s</span><span class="p">))</span>

<span class="c1">#Another example:</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1st&quot;</span><span class="p">,</span><span class="s2">&quot;2nd&quot;</span><span class="p">,</span><span class="s2">&quot;3rd&quot;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cont_handle</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">,</span> <span class="n">seq</span>  <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">SimpleFastaParser</span><span class="p">(</span><span class="n">cont_handle</span><span class="p">))</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="s2">&quot;entry:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="s2">&quot; has size &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Header:
gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence
Sequence:
TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATC
AGGGGGTAATCTGCTCTCCTGATTCAGGAGAGTTTATGGTCACTTTTGAG
...


1st entry:
MDC020656.85  has size  2802
GAGGGGTTTAGTTCCTCATACTCGCAAAGCAAAGATACATAAATTTAGAA...

2nd entry:
MDC013284.379  has size  5173
TATCGTTTCCTCTGAGTAGAATATCGTTATAACAAGATTTTTTTTTTCCT...

3rd entry:
MDC018185.241  has size  23761
AAAACGAGGAAAATCCATCTTGATGAACAGGAGATGCGGAGGAAAAAAAT...

</pre></div></div>
</div>
<div class="section" id="BioSeqIO.parse()">
<h3>BioSeqIO.parse()<a class="headerlink" href="#BioSeqIO.parse()" title="Permalink to this headline">¶</a></h3>
<p>The method <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.parse</span></code> is used to parse some sequence data into a <strong>SeqRecord iterator</strong>. In particular, the basic syntax is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>SeqRecordIterator = Bio.SeqIO.parse(filename, file_format)
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">filename</span></code> is typically an open handle to a file and <code class="docutils literal notranslate"><span class="pre">file_format</span></code> is a lower case string describing the file format. Possible options include <strong>fasta, fastq-illumina, abi, ace, clustal…</strong> all the options are available <a class="reference external" href="http://biopython.org/wiki/SeqIO">here</a>.</p>
<p><strong>Example:</strong> Let’s read the first 10 entries in the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> printing off the length of the sequence and the first 50 bases of each sequence followed by “…”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">cnt</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="k">if</span><span class="p">(</span><span class="n">cnt</span> <span class="o">==</span> <span class="mi">10</span><span class="p">):</span>
        <span class="k">break</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Seq </span><span class="si">{}</span><span class="s2"> has length </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="p">)))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">cnt</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Seq MDC020656.85 has length 2802
GAGGGGTTTAGTTCCTCATACTCGCAAAGCAAAGATACATAAATTTAGAA...

Seq MDC001115.177 has length 3118
TGAATGGTGAAAATTAGCCAGAAGATCTTCTCCACACATGACATATGCAT...

Seq MDC013284.379 has length 5173
TATCGTTTCCTCTGAGTAGAATATCGTTATAACAAGATTTTTTTTTTCCT...

Seq MDC018185.243 has length 22724
CCATTAGTGACCCCCAATGCGGATTAACCAAGCACGGTCAAGATAACCAA...

Seq MDC018185.241 has length 23761
AAAACGAGGAAAATCCATCTTGATGAACAGGAGATGCGGAGGAAAAAAAT...

Seq MDC004527.213 has length 3551
CTTGTATGTTGAAGCTTTGTGAGTGGAGCATATAGGTTGAGGTAGTGTTC...

Seq MDC003661.174 has length 3334
AGTAAGGTTGTAGAATTCAATTTCCACGGAACTCCAGCACAGCTTAGGCA...

Seq MDC012176.157 has length 2236
ATAGTCACTGACAGCCGGTGACAGTTCGACAACAATAAGTTCAGGATGTT...

Seq MDC001204.812 has length 5804
TAATGTCCGAAAAAGATGAAAATGTAGTTTGCATGAAGAAGGAGCTGTCC...

Seq MDC001204.810 has length 9798
ATGCTGCTGACATATTCCAATCGACTACACGTCACATCCGTTGCTATTAT...

</pre></div></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.parse</span></code> returns an iterator, therefore it is possible to manually fetch one SeqRecord after the other with the <code class="docutils literal notranslate"><span class="pre">next(iterator)</span></code> method.</p>
<p><strong>Example:</strong> Let’s read the first 3 entries of the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> printing off the length of the sequence and the first 50 bases of each sequence followed by “…”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">seqIterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1st&quot;</span><span class="p">,</span><span class="s2">&quot;2nd&quot;</span><span class="p">,</span><span class="s2">&quot;3rd&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
    <span class="n">seqRec</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">seqIterator</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="s2">&quot;entry:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot; has size &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1st entry:
MDC020656.85  has size  2802
GAGGGGTTTAGTTCCTCATACTCGCAAAGCAAAGATACATAAATTTAGAA...

2nd entry:
MDC001115.177  has size  3118
TGAATGGTGAAAATTAGCCAGAAGATCTTCTCCACACATGACATATGCAT...

3rd entry:
MDC013284.379  has size  5173
TATCGTTTCCTCTGAGTAGAATATCGTTATAACAAGATTTTTTTTTTCCT...

</pre></div></div>
</div>
</div>
<div class="section" id="Sequences-as-dictionaries">
<h3>Sequences as dictionaries<a class="headerlink" href="#Sequences-as-dictionaries" title="Permalink to this headline">¶</a></h3>
<p>The module <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> also has three different ways to allow random access to elements:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.to_dict(file_handle/iterator)</span></code> : builds a dictionary of all the SeqRecords keeping them in memory and allowing modifications to the records. <strong>This potentially uses a lot of memory but is very fast</strong>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index(filename,file_type)</span></code> : builds a sort of read-only dictionary, parses the elements into SeqRecords on demand (i.e. it returns an iterator!). <strong>This method is slower, but more memory efficient</strong>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index_db(indexName.idx,filenames,</span> <span class="pre">file_format)</span></code> : builds a read-only dictionary, but stores ids and offsets on a SQLite3 database. <strong>It is slower but uses less memory.</strong></p></li>
</ol>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.to_dict(file_handle/iterator)</span></code> is a open file handler or the iterator to SeqRecords (like the output of Bio.SeqIO.parse()).</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> in memory printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>


</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;MDC020656.85&#39;, &#39;MDC001115.177&#39;, &#39;MDC013284.379&#39;, &#39;MDC018185.243&#39;, &#39;MDC018185.241&#39;]

Number of sequences:  82

Description: MDC019140.399

ID: MDC019140.399
Name: MDC019140.399
Description: MDC019140.399
Number of features: 0
Seq(&#39;GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT&#39;)
</pre></div></div>
</div>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index(filename,file_type)</span></code> is the file name to read sequences from and the file type (like fasta, genebank,…).</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="files_samples/contigs82.fasta">contigs82.fasta</a> in memory (as an index) printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span>  <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SeqDict is an iterator:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SeqDict is an iterator:
KeysView(SeqIO.index(&#39;file_samples/contigs82.fasta&#39;, &#39;fasta&#39;, alphabet=None, key_function=None))

[&#39;MDC020656.85&#39;, &#39;MDC001115.177&#39;, &#39;MDC013284.379&#39;, &#39;MDC018185.243&#39;, &#39;MDC018185.241&#39;, &#39;MDC004527.213&#39;, &#39;MDC003661.174&#39;, &#39;MDC012176.157&#39;, &#39;MDC001204.812&#39;, &#39;MDC001204.810&#39;, &#39;MDC004389.256&#39;, &#39;MDC024257.15&#39;, &#39;MDC018297.229&#39;, &#39;MDC001802.364&#39;, &#39;MDC016621.241&#39;, &#39;MDC014057.243&#39;, &#39;MDC021015.302&#39;, &#39;MDC018185.242&#39;, &#39;MDC051782.000&#39;, &#39;MDC017187.314&#39;, &#39;MDC017187.311&#39;, &#39;MDC012865.410&#39;, &#39;MDC000427.83&#39;, &#39;MDC017187.319&#39;, &#39;MDC017187.318&#39;, &#39;MDC004081.319&#39;, &#39;MDC021913.275&#39;, &#39;MDC015147.205&#39;, &#39;MDC000038.355&#39;, &#39;MDC016032.95&#39;, &#39;MDC052568.000&#39;, &#39;MDC008119.414&#39;, &#39;MDC026201.7&#39;, &#39;MDC003995.601&#39;, &#39;MDC009211.561&#39;, &#39;MDC009211.567&#39;, &#39;MDC054294.001&#39;, &#39;MDC004364.265&#39;, &#39;MDC002360.219&#39;, &#39;MDC003408.117&#39;, &#39;MDC015155.172&#39;, &#39;MDC053310.000&#39;, &#39;MDC019140.398&#39;, &#39;MDC019140.399&#39;, &#39;MDC011390.337&#39;, &#39;MDC007154.375&#39;, &#39;MDC006346.716&#39;, &#39;MDC010588.505&#39;, &#39;MDC002519.240&#39;, &#39;MDC031322.5&#39;, &#39;MDC010588.502&#39;, &#39;MDC006346.711&#39;, &#39;MDC011551.182&#39;, &#39;MDC002717.156&#39;, &#39;MDC006346.719&#39;, &#39;MDC007838.447&#39;, &#39;MDC007018.186&#39;, &#39;MDC017873.233&#39;, &#39;MDC016296.138&#39;, &#39;MDC019067.226&#39;, &#39;MDC036568.1&#39;, &#39;MDC014019.318&#39;, &#39;MDC007995.528&#39;, &#39;MDC026961.60&#39;, &#39;MDC013443.168&#39;, &#39;MDC022800.298&#39;, &#39;MDC021558.159&#39;, &#39;MDC002479.192&#39;, &#39;MDC010751.301&#39;, &#39;MDC000219.190&#39;, &#39;MDC020963.161&#39;, &#39;MDC020963.162&#39;, &#39;MDC051637.000&#39;, &#39;MDC005174.220&#39;, &#39;MDC040033.7&#39;, &#39;MDC019674.147&#39;, &#39;MDC010450.877&#39;, &#39;MDC007097.457&#39;, &#39;MDC016278.70&#39;, &#39;MDC013723.254&#39;, &#39;MDC002838.179&#39;, &#39;MDC009771.217&#39;]

Number of sequences:  82

Description: MDC019140.399

ID: MDC019140.399
Name: MDC019140.399
Description: MDC019140.399
Number of features: 0
Seq(&#39;GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT&#39;)
</pre></div></div>
</div>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index_db(indexName.idx,filenames,</span> <span class="pre">file_format)</span></code> is the name of the (SQLite3) index to store (.idx), the file (or files) to index and the format of those files like fasta, genebank,…</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="files_samples/contigs82.fasta">contigs82.fasta</a> in memory (as an index) printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span>  <span class="n">SeqIO</span><span class="o">.</span><span class="n">index_db</span><span class="p">(</span><span class="s2">&quot;file_samples/ctgs.idx&quot;</span><span class="p">,</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SeqDict is an iterator:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SeqDict is an iterator:
KeysView(SeqIO.index_db(&#39;file_samples/ctgs.idx&#39;, filenames=[&#39;file_samples/contigs82.fasta&#39;], format=&#39;fasta&#39;, key_function=None))

[&#39;MDC020656.85&#39;, &#39;MDC001115.177&#39;, &#39;MDC013284.379&#39;, &#39;MDC018185.243&#39;, &#39;MDC018185.241&#39;, &#39;MDC004527.213&#39;, &#39;MDC003661.174&#39;, &#39;MDC012176.157&#39;, &#39;MDC001204.812&#39;, &#39;MDC001204.810&#39;, &#39;MDC004389.256&#39;, &#39;MDC024257.15&#39;, &#39;MDC018297.229&#39;, &#39;MDC001802.364&#39;, &#39;MDC016621.241&#39;, &#39;MDC014057.243&#39;, &#39;MDC021015.302&#39;, &#39;MDC018185.242&#39;, &#39;MDC051782.000&#39;, &#39;MDC017187.314&#39;, &#39;MDC017187.311&#39;, &#39;MDC012865.410&#39;, &#39;MDC000427.83&#39;, &#39;MDC017187.319&#39;, &#39;MDC017187.318&#39;, &#39;MDC004081.319&#39;, &#39;MDC021913.275&#39;, &#39;MDC015147.205&#39;, &#39;MDC000038.355&#39;, &#39;MDC016032.95&#39;, &#39;MDC052568.000&#39;, &#39;MDC008119.414&#39;, &#39;MDC026201.7&#39;, &#39;MDC003995.601&#39;, &#39;MDC009211.561&#39;, &#39;MDC009211.567&#39;, &#39;MDC054294.001&#39;, &#39;MDC004364.265&#39;, &#39;MDC002360.219&#39;, &#39;MDC003408.117&#39;, &#39;MDC015155.172&#39;, &#39;MDC053310.000&#39;, &#39;MDC019140.398&#39;, &#39;MDC019140.399&#39;, &#39;MDC011390.337&#39;, &#39;MDC007154.375&#39;, &#39;MDC006346.716&#39;, &#39;MDC010588.505&#39;, &#39;MDC002519.240&#39;, &#39;MDC031322.5&#39;, &#39;MDC010588.502&#39;, &#39;MDC006346.711&#39;, &#39;MDC011551.182&#39;, &#39;MDC002717.156&#39;, &#39;MDC006346.719&#39;, &#39;MDC007838.447&#39;, &#39;MDC007018.186&#39;, &#39;MDC017873.233&#39;, &#39;MDC016296.138&#39;, &#39;MDC019067.226&#39;, &#39;MDC036568.1&#39;, &#39;MDC014019.318&#39;, &#39;MDC007995.528&#39;, &#39;MDC026961.60&#39;, &#39;MDC013443.168&#39;, &#39;MDC022800.298&#39;, &#39;MDC021558.159&#39;, &#39;MDC002479.192&#39;, &#39;MDC010751.301&#39;, &#39;MDC000219.190&#39;, &#39;MDC020963.161&#39;, &#39;MDC020963.162&#39;, &#39;MDC051637.000&#39;, &#39;MDC005174.220&#39;, &#39;MDC040033.7&#39;, &#39;MDC019674.147&#39;, &#39;MDC010450.877&#39;, &#39;MDC007097.457&#39;, &#39;MDC016278.70&#39;, &#39;MDC013723.254&#39;, &#39;MDC002838.179&#39;, &#39;MDC009771.217&#39;]

Number of sequences:  82

Description: MDC019140.399

ID: MDC019140.399
Name: MDC019140.399
Description: MDC019140.399
Number of features: 0
Seq(&#39;GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT&#39;)
</pre></div></div>
</div>
<p>Note that a file ctgs.idx has been created to store ids and offsets.</p>
</div>
<div class="section" id="Writing-sequence-files">
<h3>Writing sequence files<a class="headerlink" href="#Writing-sequence-files" title="Permalink to this headline">¶</a></h3>
<p>SeqRecords can be written out to files by using</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>N = Bio.SeqIO.write(records,out_filename, file_format)
</pre></div>
</div>
<p>where <strong>records</strong> is a list of the SeqRecords to write, <strong>out_filename</strong> is the string with the filename to write and <strong>file_format</strong> is the format of the file to write. <strong>N</strong> is the number of sequences written.</p>
<div class="admonition warning">
<p><strong>WARNING:</strong> If you write a file that is already present, <code class="docutils literal notranslate"><span class="pre">SeqIO.write</span></code> will just rewrite it without telling you.</p>
</div>
<p><strong>Example:</strong> Write three protein records to file then read the file and print the three sequences.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="n">rec1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
<span class="o">+</span><span class="s2">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
<span class="o">+</span><span class="s2">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
<span class="o">+</span><span class="s2">&quot;SSAC&quot;</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="p">)</span>

<span class="n">rec2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ&quot;</span> \
<span class="o">+</span><span class="s2">&quot;DMVVVEIPKLGKEAAVKAIKEWGQ&quot;</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|13919613|gb|AAK33142.1|&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Fragaria vesca subsp. bracteata]&quot;</span><span class="p">)</span>

<span class="n">rec3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC&quot;</span> \
<span class="o">+</span><span class="s2">&quot;EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP&quot;</span> \
<span class="o">+</span><span class="s2">&quot;KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN&quot;</span> \
<span class="o">+</span><span class="s2">&quot;NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV&quot;</span> \
<span class="o">+</span><span class="s2">&quot;SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW&quot;</span> \
<span class="o">+</span><span class="s2">&quot;IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT&quot;</span> \
<span class="o">+</span><span class="s2">&quot;TGEGLEWGVLFGFGPGLTVETVVLHSVAT&quot;</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|13925890|gb|AAK49457.1|&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Nicotiana tabacum]&quot;</span><span class="p">)</span>

<span class="n">my_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">rec1</span><span class="p">,</span> <span class="n">rec2</span><span class="p">,</span> <span class="n">rec3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rec1</span><span class="p">)</span>

<span class="n">seqWritten</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">my_records</span><span class="p">,</span> <span class="s2">&quot;file_samples/my_proteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqWritten</span><span class="p">,</span> <span class="s2">&quot;sequences written!&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reading file&quot;</span><span class="p">)</span>
<span class="n">seqIT</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/my_proteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="n">cnt</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqIT</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence &quot;</span><span class="p">,</span> <span class="n">cnt</span><span class="p">,</span> <span class="s2">&quot;(&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">cnt</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ID: gi|14150838|gb|AAK54648.1|AF376133_1
Name: &lt;unknown name&gt;
Description: chalcone synthase [Cucumis sativus]
Number of features: 0
Seq(&#39;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVG...SAC&#39;)
3 sequences written!

Reading file
Sequence  1 ( gi|14150838|gb|AAK54648.1|AF376133_1 )
MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGDGAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNMSSAC

Sequence  2 ( gi|13919613|gb|AAK33142.1| )
YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQDMVVVEIPKLGKEAAVKAIKEWGQ

Sequence  3 ( gi|13925890|gb|AAK49457.1| )
MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMCEKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQPKSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAENNKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELVSAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFWIAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGTTGEGLEWGVLFGFGPGLTVETVVLHSVAT

</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Multiple-sequence-alignment">
<h2>Multiple sequence alignment<a class="headerlink" href="#Multiple-sequence-alignment" title="Permalink to this headline">¶</a></h2>
<p><strong>Multiple Sequence Alignments</strong>, as the name says, are a collection of multiple sequences which have been aligned together – usually with the insertion of gap characters, and addition of leading or trailing gaps – such that all the sequence strings have the same length.</p>
<img alt="msa" src="_images/msa.gif" />
<p>Such an alignment can be regarded as a matrix of letters, in biopython each row is a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object. These alignments are stored in an object <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code></p>
<div class="section" id="Parsing-multiple-sequence-alignments">
<h3>Parsing multiple sequence alignments<a class="headerlink" href="#Parsing-multiple-sequence-alignments" title="Permalink to this headline">¶</a></h3>
<p>The function <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.parse()</span></code> returns an iterator of <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code> objects. In the frequent case that we have to deal with a single alignment we will have to use the <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.read()</span></code> function.</p>
<p>The basic syntax of the two functions:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.parse(file_handle,</span> <span class="pre">alignment_format)</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.read(file_handle,</span> <span class="pre">alignment_format)</span></code></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">file_handle</span></code> is the handler to the opened file, while the <code class="docutils literal notranslate"><span class="pre">alignment_format</span></code> is a lower case string with the alignment format (e.g. fasta, clustal, stockholm, mauve, phylip,…). Check <a class="reference external" href="http://biopython.org/wiki/AlignIO">here</a> for all possible formats.</p>
<p>When more than one multiple alignment has to be read we will have to use the <code class="docutils literal notranslate"><span class="pre">parse</span></code> function, for single entries, we would use the function <code class="docutils literal notranslate"><span class="pre">read</span></code>.</p>
<p><strong>Example:</strong> Load and visualize the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format. Information on the format can be found <a class="reference external" href="https://en.wikipedia.org/wiki/Stockholm_format">here</a>. The file is available here: <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Alignment with 16 rows and 395 columns
YLFFILDK-NSPEP-YGSIKRVCNTMLGVPSQCAISKHILQS--...QDV AGO1_SCHPO/500-799
FILCILPERKTSDI-YGPWKKICLTEEGIHTQCICPIKI-----...FTK AGO6_ARATH/541-851
FILCVLPDKKNSDL-YGPWKKKNLTEFGIVTQCMAPTRQPND--...FMK AGO4_ARATH/577-885
CIIVVLQS-KNSDI-YMTVKEQSDIVHGIMSQCVLMKNVSRP--...HVK TAG76_CAEEL/660-966
LIVVVLPG--KTPI-YAEVKRVGDTVLGIATQCVQAKNAIRT--...HLV O16720_CAEEL/566-867
TFVFIITD-DSITT-LHQRYKMIEKDTKMIVQDMKLSKALSV--...LWN O62275_CAEEL/594-924
DILVGIAR-EKKPD-VHDILKYFEESIGLQTIQLCQQTVDKMMG...NYK YQ53_CAEEL/650-977
TIVFGIIA-EKRPD-MHDILKYFEEKLGQQTIQISSETADKFMR...NYK NRDE3_CAEEL/673-1001
MLVVMLAD-DNKTR-YDSLKKYLCVECPIPNQCVNLRTLAGKSK...SLH Q17567_CAEEL/397-708
IVMVVMRS-PNEEK-YSCIKKRTCVDRPVPSQVVTLKVIAPRQQ...SIN AUB_DROME/555-852
LILCLVPN-DNAER-YSSIKKRGYVDRAVPTQVVTLKTTKNRSL...NLH PIWI_DROME/538-829
IVVCLLSS-NRKDK-YDAIKKYLCTDCPTPSQCVVARTLGKQQT...SIH PIWL1_HUMAN/555-847
GIMLVLPE-YNTPL-YYKLKSYLINS--IPSQFMRYDILSNRNL...VNR PIWI_ARCFU/110-406
CFALIIGKEKYKDNDYYEILKKQLFDLKIISQNILWENWRKDDK...NWK Y1321_METJA/426-699
LVIVFLEEYPKVDP-YKSFLLYDFVKRELLKKMIPSQVILNRTL...GIE O67434_AQUAE/419-694
LLLAILPD-NNGSL-YGDLKRICETELGLISQCCLTKHVFKI--...YLE AGO10_ARATH/625-946
</pre></div></div>
</div>
<p>The code above prints a summary of the information, but each <code class="docutils literal notranslate"><span class="pre">Bio.Align.MultipleSeqAlignment</span></code> is composed of several alignment records (<code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>) that we can access with all their information looping through the object. Each <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> contains several information like the <strong>ID, Name, Description, Number of features, start, end and sequence</strong>.</p>
<p><strong>Example:</strong></p>
<p>Load the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>. For each record print the description, start and the end point, the sequence and the external dbxrefs (if any).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>


<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">description</span>
    <span class="n">dbref</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">align</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> S:</span><span class="si">{}</span><span class="s2"> E:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">desc</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dbref</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">dbref</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AGO1_SCHPO/500-799 S:500 E:799
YLFFILDK-NSPEP-YGSIKRVCNTMLGVPSQCAISKHILQS---------KPQYCANLGMKINVKVGGIN-CSLIPKSNP----LGNVPTL---------ILGGDVYHPGVGA----------TGVSIASIVASVD-LNGCKYTAVSRSQPRHQEVIEG-MKD------------IVVYLLQGFRAMTKQ-QPQRIIYFRDGTSEGQFLSVINDELSQIKEACH-------SLSPKYN--PKILVCTTQKRHHARFFIKNKSDG----------------------DRNGNPLPGTII---EKHVTHPYQYDFYLISHPSLQGVSVPVHYTVLHDEIQMPPDQF-QTL------CYNLCYVYARAT----SAVSLVPPVYYAHLVSNLARYQDV

AGO6_ARATH/541-851 S:541 E:851
FILCILPERKTSDI-YGPWKKICLTEEGIHTQCICPIKI------------SDQYLTNVLLKINSKLGGIN-SLLGIEYSYNIPLINKIPTL---------ILGMDVSHGPPGR---------ADVPSVAAVVGSKCWPLISRYRAAVRTQSPRLEMIDSLFQPIENTE--KGDNGIMNELFVEFYRTSRARKPKQIIIFRDGVSESQFEQVLKIEVDQIIKAYQ-------RLGESDV--PKFTVIVAQKNHHTKLFQAKGPE---------------------------NVPAGTVV---DTKIVHPTNYDFYMCAHAGKIGTSRPAHYHVLLDEIGFSPDDL-QNL------IHSLSYVNQRST----TATSIVAPVRYAHLAAAQVAQFTK

AGO4_ARATH/577-885 S:577 E:885
FILCVLPDKKNSDL-YGPWKKKNLTEFGIVTQCMAPTRQPND-----------QYLTNLLLKINAKLGGLN-SMLSVERTPAFTVISKVPTI---------ILGMDVSHGSPGQ---------SDVPSIAAVVSSREWPLISKYRASVRTQPSKAEMIESLVKKNGT-----EDDGIIKELLVDFYTSSNKRKPEHIIIFRDGVSESQFNQVLNIELDQIIEACK-------LLDANWN--PKFLLLVAQKNHHTKFFQPTSPE---------------------------NVPPGTII---DNKICHPKNNDFYLCAHAGMIGTTRPTHYHVLYDEIGFSADEL-QEL------VHSLSYVYQRST----SAISVVAPICYAHLAAAQLGTFMK

TAG76_CAEEL/660-966 S:660 E:966
CIIVVLQS-KNSDI-YMTVKEQSDIVHGIMSQCVLMKNVSRP---------TPATCANIVLKLNMKMGGIN--SRIVADKITNKYLVDQPTM---------VVGIDVTHPTQAEM-------RMNMPSVAAIVANVD-LLPQSYGANVKVQKKCRESVVY-LLD------------AIRERIITFYRHTKQ-KPARIIVYRDGVSEGQFSEVLREEIQSIRTACL-------AIAEDFR--PPITYIVVQKRHHARIFCKYQNDM---------------------VGKAKNVPPGTTV---DTGIVSPEGFDFYLCSHYGVQGTSRPARYHVLLDECKFTADEI-QSI------TYGMCHTYGRCT----RSVSIPTPVYYADLVATRARCHVK

O16720_CAEEL/566-867 S:566 E:867
LIVVVLPG--KTPI-YAEVKRVGDTVLGIATQCVQAKNAIRT---------TPQTLSNLCLKMNVKLGGVN-SILLPNVRPR---IFNEPVI---------FLGCDITHPAAGD---------TRKPSIAAVVGSMD-AHPSRYAATVRVQQHRQEIITD-LTY------------MVRELLVQFYRNTRF-KPARIVVYRDGVSEGQLFNVLQYELRAIREACV-------MLESGYQ--PGITFIAVQKRHHTRLFAADKADQ---------------------VGKAFNIPPGTTV---DVGITHPTEFDFFLCSHAGIQGTSRPSHYHVLWDDNDLTADEL-QQL------TYQMCHTYVRCT----RSVSIPAPAYYAHLVAFRARYHLV

O62275_CAEEL/594-924 S:594 E:924
TFVFIITD-DSITT-LHQRYKMIEKDTKMIVQDMKLSKALSV--IN---AGKRLTLENVINKTNVKLGGSN--YVFVDAKKQL-----DSHL---------IIGVGISAPPAGTKYAM-ENKGVLNPNVIGYAYNA--QHNQEFSGDFVLNSASQDTLAP-IED------------IVMHSLNEYQKFHDGGLPRRVIVYRTGTSEGNHGSIMAYEIPLARAAMR-------DFSPDIQ----LVYIVVSKDHSFRFFKPDLASLASRPQATSSTASRHSAMPAAPKAWDLNIAPGILV---DSIVTNPACKQFFLNSHITLQGTAKTPLYTVLADDAKVSMTAL-EDI------TYKLCHLHQIVG----LPTSLPTPLYVANEYAKRGRNLWN

YQ53_CAEEL/650-977 S:650 E:977
DILVGIAR-EKKPD-VHDILKYFEESIGLQTIQLCQQTVDKMMGG----QGGRQTIDNVMRKFNLKCGGTNFFVEIPNAVRGKAVCSNNETLRKKLLEHVQFIGFEISHGASRTLFDRSRSQMDGEPSVVGVSYSLT---NSTQLGGFTYLQTQKEYKLQKLDE------------FFPKCVRSYKEHSKT-LPTRIVIYRVGAGEGNFNRVKE-EVEEMRRTFD-------KIQPGYR--PHLVVIIAQRASHARVFPSCISGN---------------------RATDQNIPSGTCV---ENVLTSYGYDEFILSSQTPLIGTVRPCKYTILVNDAKWSKNEL-MHL------TYFRAFGHQVSY----QPPSVPDVLYAAENLAKRGRNNYK

NRDE3_CAEEL/673-1001 S:673 E:1001
TIVFGIIA-EKRPD-MHDILKYFEEKLGQQTIQISSETADKFMRD----HGGKQTIDNVIRKLNPKCGGTNFLIDVPESVGHRVVCNNSAEMRAKLYAKTQFIGFEMSHTGARTRFDIQKVMFDGDPTVVGVAYSL--KHSAQLGGFSYFQESRLHKLTN-LQE------------KMQICLNAYEQSSSY-LPETVVVYRVGSGEGDYPQIVN-EVNEMKLAAR-------KKKHGYN--PKFLVICTQRNSHIRVFPEHINERG--------------------KSMEQNVKSGTCV---DVPGASHGYEEFILCCQTPLIGTVKPTKYTIIVNDCRWSKNEI-MNV------TYHLAFAHQVSY----APPAIPNVSYAAQNLAKRGHNNYK

Q17567_CAEEL/397-708 S:397 E:708
MLVVMLAD-DNKTR-YDSLKKYLCVECPIPNQCVNLRTLAGKSKDGGENKNLGSIVLKIVLQMICKTGGA---LWKVNIPL-------KSTM---------IVGYDLYHDSTLK-----------GKTVGACVSTTS-NDFTQFYSQTRPHENPTQLGNN-LTH------------FVRKSLKQYYDNNDKTLPSRLILYRDGAGDGQIPYIKNTEVKLVRDACDAVTDKAAELSNKVQEKIKLAFIIVTKRVNMRILKQGSSS-----------------------KSAINPQPGTVV---DTTVTRPERMDFYLVPQFVNQGTVTPVSYNIIHDDTGLGPDKH-QQL------AFKLCHLYYNWQ----GTVRVPAPCQYAHKLAFLTAQSLH

AUB_DROME/555-852 S:555 E:852
IVMVVMRS-PNEEK-YSCIKKRTCVDRPVPSQVVTLKVIAPRQQKP---TGLMSIATKVVIQMNAKLMGA---PWQVVIPL-------HGLM---------TVGFDVCHSPKNK-----------NKSYGAFVATMDQKESFRYFSTVNEHIKGQELSEQ-MSV------------NMACALRSYQEQHRS-LPERILFFRDGVGDGQLYQVVNSEVNTLKDRLDEI---YKSAGKQEG--CRMTFIIVSKRINSRYFT-----------------------------GHRNPVPGTVV---DDVITLPERYDFFLVSQAVRIGTVSPTSYNVISDNMGLNADKL-QML------SYKMTHMYYNYS----GTIRVPAVCHYAHKLAFLVAESIN

PIWI_DROME/538-829 S:538 E:829
LILCLVPN-DNAER-YSSIKKRGYVDRAVPTQVVTLKTTKNRSL--------MSIATKIAIQLNCKLGYT---PWMIELPL-------SGLM---------TIGFDIAKSTRDR-----------KRAYGALIASMDLQQNSTYFSTV-TECSAFDVLANTLWP------------MIAKALRQYQHEHRK-LPSRIVFYRDGVSSGSLKQLFEFEVKDIIEKLKTE---YARVQLSP---PQLAYIVVTRSMNTRFFLN-----------------------------GQNPPPGTIV---DDVITLPERYDFYLVSQQVRQGTVSPTSYNVLYSSMGLSPEKM-QKL------TYKMCHLYYNWS----GTTRVPAVCQYAKKLATLVGTNLH

PIWL1_HUMAN/555-847 S:555 E:847
IVVCLLSS-NRKDK-YDAIKKYLCTDCPTPSQCVVARTLGKQQT-------VMAIATKIALQMNCKMGGE---LWRVDIPL-------KLVM---------IVGIDCYHDMTAG-----------RRSIAGFVASIN-EGMTRWFSRCIFQDRGQELVDG-LKV------------CLQAALRAW-NSCNEYMPSRIIVYRDGVGDGQLKTLVNYEVPQFLDCLK-------SIGRGYN--PRLTVIVVKKRVNTRFFAQSGGR-------------------------LQNPLPGTVI---DVEVTRPEWYDFFIVSQAVRSGSVSPTHYNVIYDNSGLKPDHI-QRL------TYKLCHIYYNWP----GVIRVPAPCQYAHKLAFLVGQSIH

PIWI_ARCFU/110-406 S:110 E:406
PDB; 2W42 B; 110-406;,PDB; 1YTU B; 110-406;,PDB; 2BGG B; 110-406;,PDB; 1W9H A; 110-406;,PDB; 2BGG A; 110-406;,PDB; 1YTU A; 110-406;,PDB; 2W42 A; 110-406;
GIMLVLPE-YNTPL-YYKLKSYLINS--IPSQFMRYDILSNRNL--------TFYVDNLLVQFVSKLGGK---PWILNVDPEK--------------------GSDIIIGTGAT--------RIDNVNLFCFAMVFK-KDGTMLWNEISPIVTSSEYLTY-LKS------------TIKKVVYGFKKSNPDWDVEKLTLHVSG----KRPKMKDGETKILKETVE-------ELKKQEMVSRDVKYAILHLNETHPFWVMGDPN------------------------NRFHPYEGTKVKLSSKRYLLTLLQPYLKRNGLEMVTPIKPLSVEIVSDN--WTSEEYYHNVHEILDEIYYLSKMNWRGF----RSRNLPVTVNYPKLVAGIIANVNR

Y1321_METJA/426-699 S:426 E:699
CFALIIGKEKYKDNDYYEILKKQLFDLKIISQNILWENWRKDDK--------GYMTNNLLIQIMGKLGIK---YFILDSKTPYDY----------------IMGLDTGLGIFGN------------HRVGGCTVVYDSEGKIRRIQPIETPAPGERLHLP----------------YVIEYLEN--KANIDMENKNILFLRDG-------FIQNSERNDLKEISK-------ELN------SNIEVISIRKNNKYKVFTSDYRI------------------------GSVFGNDGIFLPHKTPFGSNPVKLSTWLRFNCGNEEGLK---IN----------ESI-MQL------LYDLTKMNYSALYGEGRYLRIPAPIHYADKFVKALGKNWK

O67434_AQUAE/419-694 S:419 E:694
PDB; 1YVU A; 419-694;,PDB; 2F8S A; 419-694;,PDB; 2F8T A; 419-694;,PDB; 2F8S B; 419-694;,PDB; 2NUB A; 419-694;,PDB; 2F8T B; 419-694;
LVIVFLEEYPKVDP-YKSFLLYDFVKRELLKKMIPSQVILNRTLKN---ENLKFVLLNVAEQVLAKTGNIP--YKLKEIEGKVDA----------------FVGIDISRITRDG----------KTVNAVAFTKIFNSKGELVRYYLTSYPAFGEKLTEK----------------AIGDVFSLLEKLGF-KKGSKIVVHRDG-------RLYRDEVAAFK-----------KYGELYG--YSLELLEIIKRNNPRFFSNEKFI------------------------------KGYFYKLSEDSVILATYNQVY-------EGTHQPIKVRKVYGE--LPVEVL-CSQ------ILSLTLMNYSSF----QPIKLPATVHYSDKITKLMLRGIE

AGO10_ARATH/625-946 S:625 E:946
LLLAILPD-NNGSL-YGDLKRICETELGLISQCCLTKHVFKI---------SKQYLANVSLKINVKMGGRN-TVLVDAISCRIPLVSDIPTI---------IFGADVTHPENGE---------ESSPSIAAVVASQDWPEVTKYAGLVCAQAHRQELIQDLYKTWQDPVRGTVSGGMIRDLLISFRKATGQ-KPLRIIFYRDGVSEGQFYQVLLYELDAIRKACA-------SLEPNYQ--PPVTFIVVQKRHHTRLFANNHRDKN-------------------STDRSGNILPGTVV---DTKICHPTEFDFYLCSHAGIQGTSRPAHYHVLWDENNFTADGI-QSL------TNNLCYTYARCT----RSVSIVPPAYYAHLAAFRARFYLE

</pre></div></div>
</div>
</div>
<div class="section" id="Writing-multiple-sequence-alignments">
<h3>Writing multiple sequence alignments<a class="headerlink" href="#Writing-multiple-sequence-alignments" title="Permalink to this headline">¶</a></h3>
<p>Biopython provides a function <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.write()</span></code> to write alignments to file. The basic syntax is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>N = Bio.AlignIO.write(alignments,outfile,file_format)
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">alignments</span></code> are a MultipleSeqAlignment object with the alignments to write to the output file with name <code class="docutils literal notranslate"><span class="pre">outfile</span></code> that has format <code class="docutils literal notranslate"><span class="pre">file_format</span></code> (a low case string with the file format). N is the number of entries written to the file.</p>
<p><strong>Example:</strong> Let’s create some multiple sequence alignments and store them in <a class="reference external" href="http://www.phylo.org/tools/obsolete/phylip.html">phylip</a> format. The output file is present here <a class="reference external" href="file_samples/my_malign.phy">my_malign.phy</a>. Finally, read and print the content of the file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">align1</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGDTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">),</span>
<span class="p">])</span>
<span class="n">align2</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTCAGC-AG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Delta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GACAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Epsilon&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTCAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Zeta&quot;</span><span class="p">),</span>
<span class="p">])</span>
<span class="n">align3</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTAGTACAGCTG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Eta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTAGTACAGCT-&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Theta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;-CTACTACAGGTG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Iota&quot;</span><span class="p">),</span>
<span class="p">])</span>

<span class="n">my_alignments</span> <span class="o">=</span> <span class="p">[</span><span class="n">align1</span><span class="p">,</span> <span class="n">align2</span><span class="p">,</span> <span class="n">align3</span><span class="p">]</span>
<span class="n">N</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">my_alignments</span><span class="p">,</span> <span class="s2">&quot;file_samples/my_malign.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> entries written to the file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/my_malign.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3 entries written to the file

Alignment with 3 rows and 12 columns
ACTGCTAGCTAG Alpha
ACT-CTAGCTAG Beta
ACTGCTAGDTAG Gamma
Alignment with 3 rows and 9 columns
GTCAGC-AG Delta
GACAGCTAG Epsilon
GTCAGCTAG Zeta
Alignment with 3 rows and 13 columns
ACTAGTACAGCTG Eta
ACTAGTACAGCT- Theta
-CTACTACAGGTG Iota
</pre></div></div>
</div>
<p>The content of the file is reported in the following screenshot:</p>
<img alt="phylimg" src="_images/phylip_aligns.png" />
<p>Note that it is also possible to convert one format into the other (provided that all information needed for the second format is available) by using the function:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Bio.AlignIO.convert(input_file, input_file_format, output_file, output_file_format)
</pre></div>
</div>
<p>basically by passing the input file name and format and output file name and format.</p>
<p>Ex:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Bio.AlignIO.convert(&quot;PF05371_seed.sth&quot;, &quot;stockholm&quot;, &quot;PF05371_seed.aln&quot;, &quot;clustal&quot;)
</pre></div>
</div>
<p>converts a stockholm file into a clustal.</p>
</div>
<div class="section" id="Manipulating-alignments">
<h3>Manipulating alignments<a class="headerlink" href="#Manipulating-alignments" title="Permalink to this headline">¶</a></h3>
<p>It is possible to slice alignments using the [] operator applied on a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>. In particular:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[i,j]</span></code> returns the jth character of alignment i as a string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[:,j]</span></code> returns all the jth characters of the multiple alignment as a string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[:,i:j]</span></code> returns a MultipleSeqAlignment with the sub-alignments going for i to j (excluded)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[a:b,i:j]</span></code> similar to 3. but for alignments going from a to b (excluded) only</p></li>
</ol>
<p>One can also combine different portions of the alignment by slicing and adding the slices together with</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>SeqRecord[:,i:j] + SeqRecord[:,t:k]
</pre></div>
</div>
<p><strong>Example:</strong> Load the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>. For each record print the description, the first 20 bases and the final 20 bases of the sequence.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="n">edited_aligns</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[:,</span><span class="mi">0</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span> <span class="o">+</span> <span class="n">alignments</span><span class="p">[:,</span><span class="o">-</span><span class="mi">20</span><span class="p">:]</span>


<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">edited_aligns</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="o">.</span><span class="n">description</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AGO1_SCHPO/500-799       YLFFILDK-NSPEP-YGSIKLVPPVYYAHLVSNLARYQDV
AGO6_ARATH/541-851       FILCILPERKTSDI-YGPWKIVAPVRYAHLAAAQVAQFTK
AGO4_ARATH/577-885       FILCVLPDKKNSDL-YGPWKVVAPICYAHLAAAQLGTFMK
TAG76_CAEEL/660-966      CIIVVLQS-KNSDI-YMTVKIPTPVYYADLVATRARCHVK
O16720_CAEEL/566-867     LIVVVLPG--KTPI-YAEVKIPAPAYYAHLVAFRARYHLV
O62275_CAEEL/594-924     TFVFIITD-DSITT-LHQRYLPTPLYVANEYAKRGRNLWN
YQ53_CAEEL/650-977       DILVGIAR-EKKPD-VHDILVPDVLYAAENLAKRGRNNYK
NRDE3_CAEEL/673-1001     TIVFGIIA-EKRPD-MHDILIPNVSYAAQNLAKRGHNNYK
Q17567_CAEEL/397-708     MLVVMLAD-DNKTR-YDSLKVPAPCQYAHKLAFLTAQSLH
AUB_DROME/555-852        IVMVVMRS-PNEEK-YSCIKVPAVCHYAHKLAFLVAESIN
PIWI_DROME/538-829       LILCLVPN-DNAER-YSSIKVPAVCQYAKKLATLVGTNLH
PIWL1_HUMAN/555-847      IVVCLLSS-NRKDK-YDAIKVPAPCQYAHKLAFLVGQSIH
PIWI_ARCFU/110-406       GIMLVLPE-YNTPL-YYKLKLPVTVNYPKLVAGIIANVNR
Y1321_METJA/426-699      CFALIIGKEKYKDNDYYEILIPAPIHYADKFVKALGKNWK
O67434_AQUAE/419-694     LVIVFLEEYPKVDP-YKSFLLPATVHYSDKITKLMLRGIE
AGO10_ARATH/625-946      LLLAILPD-NNGSL-YGDLKIVPPAYYAHLAAFRARFYLE
</pre></div></div>
</div>
<p><strong>Example:</strong> Convert the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a> into phylip format. Print some stats on the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="n">out</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;stockholm&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;file_samples/PF02171_seed.aln&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;N. of seq: </span><span class="si">{}</span><span class="se">\n</span><span class="s2">Len of seq: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                                <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span>
                                                <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> multiple alignments converted to phylip&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
N. of seq: 16
Len of seq: 395
1 multiple alignments converted to phylip
</pre></div></div>
</div>
</div>
<div class="section" id="Biopython’s-pairwise2-alignment">
<h3>Biopython’s pairwise2 alignment<a class="headerlink" href="#Biopython’s-pairwise2-alignment" title="Permalink to this headline">¶</a></h3>
<p>Biopython supports running some classic multiple alignment tools like <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/0378111988903307?via%3Dihub">clustal</a>, <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC390337/">MUSCLE</a> but these require to be installed in your system first.</p>
<p>Biopython has its own module to make pairwise alignment. It provides two algorithms: <a class="reference external" href="https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm">Smith-Waterman</a> for local alignment and <a class="reference external" href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">Needleman-Wunsch</a> for global alignment. These methods are implemented in two Biopython functions of the <code class="docutils literal notranslate"><span class="pre">Bio.pairwise2</span></code> module:</p>
<p><code class="docutils literal notranslate"><span class="pre">pairwise2.align.globalxx()</span></code> and <code class="docutils literal notranslate"><span class="pre">pairwise2.align.localxx()</span></code>.</p>
<p>The basic syntax of the two methods is the same:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>aligns = pairwise2.align.globalxx(seq1,seq2)
aligns = pairwise2.align.localxx(seq1,seq2)
</pre></div>
</div>
<p>where seq1 and seq2 are two <code class="docutils literal notranslate"><span class="pre">str</span></code> objects. These methods return a list of alignments (at least one) that have the same <strong>optimal score</strong>. Each alignment is represented as tuples with the following 5 elements in order:</p>
<ol class="arabic simple">
<li><p>The alignment of the first sequence;</p></li>
<li><p>The alignment of the second sequence;</p></li>
<li><p>The alignment score;</p></li>
<li><p>The start of the alignment (for global alignments this is always 0);</p></li>
<li><p>The end of the alignment (for global alignments this is always the length of the alignment).</p></li>
</ol>
<p>The <strong>xx</strong> in <code class="docutils literal notranslate"><span class="pre">globalxx</span></code> or <code class="docutils literal notranslate"><span class="pre">localxx</span></code> encodes the penalties and scores for gaps and matches (and mismatches) of the alignment. The first letter is <strong>the score for a match</strong> the second letter is <strong>the penalty for a gap</strong>:</p>
<p>Match parameters can be:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">x</span></code> : means that a match scores 1 a mismatch 0;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">m</span></code> : the match and mismatch score are passed as additional params after the sequence (es. <code class="docutils literal notranslate"><span class="pre">aligns</span> <span class="pre">=</span> <span class="pre">pairwise2.align.globalmx(seq1,seq2,</span> <span class="pre">1,</span> <span class="pre">-1)</span></code> to set 1 as match score and -1 as mismatch penalty.</p></li>
</ul>
<p>Gap parameters can be:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">x</span></code> : gap penalty is 0;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">s</span></code> : same gap open and gap extend penalties for the 2 sequences (passed as additional params after seqs).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">d</span></code> : different gap open and gap extend penalties for the 2 seqs (additional params after the seqs).</p></li>
</ul>
<p><strong>Example:</strong></p>
<p>Let’s perform the alignment of the two sequences “ACCGTTATATAGGCCA” and “ACGTACTAGTATAGGCCA”</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>


<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="s2">&quot;ACCGTTATATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;ACGTACTAGTATAGGCCA&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Looping through aligns&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                                     <span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                                     <span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalms</span><span class="p">(</span><span class="s2">&quot;ACCGTTATATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;ACGTACTAGTATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Match: 1, Mismatch: -1, Gap open: -0.5, Gap extend: -0.2&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                                 <span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                                 <span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Alignment(seqA=&#39;ACCGT--TA-TATAGGCCA&#39;, seqB=&#39;A-CGTACTAGTATAGGCCA&#39;, score=15.0, start=0, end=19)
Alignment(seqA=&#39;ACCGT--TA-TATAGGCCA&#39;, seqB=&#39;AC-GTACTAGTATAGGCCA&#39;, score=15.0, start=0, end=19)

Looping through aligns
ACCGT--TA-TATAGGCCA
A-CGTACTAGTATAGGCCA
Score: 15.0, Start: 0, End: 19

ACCGT--TA-TATAGGCCA
AC-GTACTAGTATAGGCCA
Score: 15.0, Start: 0, End: 19


Match: 1, Mismatch: -1, Gap open: -0.5, Gap extend: -0.2
ACCGT--TA-TATAGGCCA
A-CGTACTAGTATAGGCCA
Score: 13.3, Start: 0, End: 19

ACCGT--TA-TATAGGCCA
AC-GTACTAGTATAGGCCA
Score: 13.3, Start: 0, End: 19

</pre></div></div>
</div>
<p>It is also possible to specify a substitution matrix like the <a class="reference external" href="https://en.wikipedia.org/wiki/BLOSUM">BLOSUM</a> to improve the alignment of protein sequences. Substitution matrices like the <strong>blosum62</strong> are present in Biopython and before using it we need to import it with <code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">Bio.SubsMat.MatrixInfo</span> <span class="pre">import</span> <span class="pre">blosum62</span></code>.</p>
<p>You can find all the substitution matrices available <a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SubsMat.MatrixInfo.html">here</a>.</p>
<p><strong>Example:</strong> Align the alpha and beta subunits of the human hemoglobin protein stored in <a class="reference external" href="file_samples/hem_alpha.fasta">hem_alpha.fasta</a> and <a class="reference external" href="file_samples/hem_beta.fasta">hem_beta.fasta</a> with globalxx and with a blosum62 substitution matrix and gap open -10 and gap extension -0.5.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.pairwise2</span> <span class="kn">import</span> <span class="n">format_alignment</span>
<span class="kn">from</span> <span class="nn">Bio.SubsMat.MatrixInfo</span> <span class="kn">import</span> <span class="n">blosum62</span>



<span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/hem_alpha.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/hem_beta.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>


<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span><span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span> <span class="s2">&quot;optimal alignments&quot;</span><span class="p">)</span>

<span class="n">align</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalds</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">blosum62</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span> <span class="s2">&quot;optimal alignments&quot;</span><span class="p">)</span>

<span class="n">align</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
80 optimal alignments
MV-LSPADKTNV---K-A--A-WGKVGAHAG---EY-GA-EALE-RMFLSF----PTTK-TY--FPHFDL-SH-G--S---AQVK-G------HGKKV--A--DA-LTNAVAHVD-DMPNALS----A-LSD-LHAH--KLR-VDPV-NFKL-LSHCL---LVT--LAAHLPA----EFTPA-VH-ASLDKFLAS---VSTV------LTS--KYR-
MVHL-----T--PEEKSAVTALWGKV-----NVDE-VG-GEAL-GR--L--LVVYP---WT-QRF--F--ES-FGDLSTPDA-V-MGNPKVKAHGKKVLGAFSD-GL----AH--LD--N-L-KGTFATLS-EL--HCDKL-HVDP-ENF--RL---LGNVLV-CVL-AH---HFGKEFTP-PV-QA------A-YQKV--VAGVANAL--AHKY-H
Score: 72.0, Start: 0, End: 217

2 optimal alignments
MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-----HGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR
MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH
Score: 292.5, Start: 0, End: 149
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Exercises">
<h2>Exercises<a class="headerlink" href="#Exercises" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>Write a python function that reads a genebank file given in input and prints off the following information:</p>
<ol class="arabic simple">
<li><p>Identifier, name and description;</p></li>
<li><p>The first 100 characters of the sequence;</p></li>
<li><p>Number of external references (dbxrefs) and ids of the external refs.</p></li>
<li><p>The name of the organism (hint: check the annotations dictionary at the key “organism”)</p></li>
<li><p>Retrieve and print all (if any) associated publications (hint: annotation dictionary, key:“references”)</p></li>
<li><p>Retrieve and print all the locations of “CDS” features of the sequence (hint: check the features )</p></li>
</ol>
</li>
</ol>
<p>Hint: go back and check the details of the <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object.</p>
<p>Test the program downloading some files from genebank like <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NC_005816.1/">this</a> or <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/AB026718.1">this</a>. To generate the file, please, click on the link <code class="docutils literal notranslate"><span class="pre">Send</span> <span class="pre">to:</span></code> on the top right, and then select: <code class="docutils literal notranslate"><span class="pre">Complete</span> <span class="pre">Record,</span> <span class="pre">File,</span> <span class="pre">Format:</span> <span class="pre">GeneBank</span> <span class="pre">(full),</span> <span class="pre">Create</span> <span class="pre">File</span></code>.</p>
<div class="tggle" onclick="toggleVisibility(&#39;ex1&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex1" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">reset</span> -s -f

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="k">def</span> <span class="nf">getGenBankInfo</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
    <span class="c1">#print(record.annotations.keys())</span>
    <span class="c1">#print(record.annotations)</span>

    <span class="n">ident</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">description</span>
    <span class="n">numDbrefs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">)</span>
    <span class="n">dbrefs</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">)</span>
    <span class="n">organism</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span>
    <span class="n">pubs</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ID:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ident</span><span class="p">)</span> <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;NAME:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;DESC:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">desc</span><span class="p">)</span> <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ORGANISM:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">organism</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;CDS features:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;num DBrefs </span><span class="si">{}</span><span class="s2">:&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">numDbrefs</span><span class="p">))</span>
    <span class="k">if</span><span class="p">(</span><span class="n">numDbrefs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Refs:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dbrefs</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pubs</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;PUBLICATIONS:&quot;</span><span class="p">)</span>
        <span class="n">pcount</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pubs</span><span class="p">:</span>
            <span class="c1">#print(p)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;REFERENCE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pcount</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">title</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">authors</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">journal</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">pcount</span> <span class="o">+=</span> <span class="mi">1</span>



<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;file_samples/NC_005816.gb&quot;</span>
<span class="n">getGenBankInfo</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ID:NC_005816.1
NAME:NC_005816
DESC:Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence
ORGANISM:Yersinia pestis biovar Microtus str. 91001

CDS features:
CDS [86:1109](+)
CDS [1105:1888](+)
CDS [2924:3119](+)
CDS [&lt;3358:3442](+)
CDS [4354:4780](+)
CDS [4814:5888](-)
CDS [6004:6421](+)
CDS [6663:7602](+)
CDS [7788:8088](-)
CDS [8087:8435](-)
CDS [8507:8639](-)
CDS [9264:9558](-)
num DBrefs 3:
Refs:BioProject:PRJNA224116,BioSample:SAMN02602970,Assembly:GCF_000007885.1

PUBLICATIONS:
REFERENCE [1]
Genetics of metabolic variations between Yersinia pestis biovars and the proposal of a new biovar, microtus
Zhou,D., Tong,Z., Song,Y., Han,Y., Pei,D., Pang,X., Zhai,J., Li,M., Cui,B., Qi,Z., Jin,L., Dai,R., Du,Z., Wang,J., Guo,Z., Wang,J., Huang,P. and Yang,R.
J. Bacteriol. 186 (15), 5147-5152 (2004)

REFERENCE [2]
Complete genome sequence of Yersinia pestis strain 91001, an isolate avirulent to humans
Song,Y., Tong,Z., Wang,J., Wang,L., Guo,Z., Han,Y., Zhang,J., Pei,D., Zhou,D., Qin,H., Pang,X., Han,Y., Zhai,J., Li,M., Cui,B., Qi,Z., Jin,L., Dai,R., Chen,F., Li,S., Ye,C., Du,Z., Lin,W., Wang,J., Yu,J., Yang,H., Wang,J., Huang,P. and Yang,R.
DNA Res. 11 (3), 179-197 (2004)

REFERENCE [3]
Direct Submission
Song,Y., Tong,Z., Wang,L., Han,Y., Zhang,J., Pei,D., Wang,J., Zhou,D., Han,Y., Pang,X., Zhai,J., Chen,F., Qin,H., Wang,J., Li,S., Guo,Z., Ye,C., Du,Z., Lin,W., Wang,J., Yu,J., Yang,H., Wang,J., Huang,P. and Yang,R.
Submitted (24-APR-2003) The Institute of Microbiology and Epidemiology, Academy of Military Medical Sciences, No. 20, Dongdajie Street, Fengtai District, Beijing 100071, People&#39;s Republic of China

</pre></div></div>
</div>
</div><ol class="arabic simple" start="2">
<li><p>Write a python program that loads a pfam file (stockholm format .sth) and reports for each record of the alignment:</p>
<ol class="arabic simple">
<li><p>the id of the entry</p></li>
<li><p>the start and end points</p></li>
<li><p>the number of gaps and the % of gaps on the total length of the alignment</p></li>
<li><p>the number of external database references (dbxrefs), and the first 3 external references comma separated (hint: use join).</p></li>
</ol>
</li>
</ol>
<p>Print these information to the screen. Finally, write this information in a tab separated file (.tsv) having the following format: <code class="docutils literal notranslate"><span class="pre">#ID\tstart\tend\tnum_gaps\tpercentage_gaps\tdbxrefs</span></code>.</p>
<p>You can test the program using the <a class="reference external" href="https://en.wikipedia.org/wiki/TP53">TP53</a> family file <a class="reference external" href="file_samples/PF00870_seed.sth">PF00870_seed.sth</a>.</p>
<div class="tggle" onclick="toggleVisibility(&#39;ex2&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex2" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">reset</span>  -s -f

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="k">def</span> <span class="nf">getStockholmInfo</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>

    <span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
    <span class="n">retVals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
        <span class="n">ident</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">id</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="n">gaps</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="c1">#percentage gaps</span>
        <span class="n">percGaps</span> <span class="o">=</span> <span class="mi">100</span><span class="o">*</span><span class="n">gaps</span><span class="o">/</span><span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>
        <span class="n">dbxrfsN</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">)</span>
        <span class="c1">#get top five external refs</span>
        <span class="n">dbxrfs</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">align</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ID: </span><span class="si">{}</span><span class="s2"> start: </span><span class="si">{}</span><span class="s2"> end: </span><span class="si">{}</span><span class="s2"> gaps: </span><span class="si">{}</span><span class="s2"> </span><span class="si">% g</span><span class="s2">aps: </span><span class="si">{:.2f}</span><span class="s2"> refs: </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ident</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gaps</span><span class="p">,</span> <span class="n">percGaps</span><span class="p">,</span> <span class="n">dbxrfsN</span><span class="p">,</span> <span class="n">dbxrfs</span><span class="p">))</span>
        <span class="n">retVals</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">ident</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gaps</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">percGaps</span><span class="p">),</span> <span class="n">dbxrfsN</span><span class="p">,</span> <span class="n">dbxrfs</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">retVals</span>

<span class="k">def</span> <span class="nf">writeStockholmInfo</span><span class="p">(</span><span class="n">outFile</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outFile</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#ID</span><span class="se">\t</span><span class="s2">start</span><span class="se">\t</span><span class="s2">end</span><span class="se">\t</span><span class="s2">num_gaps</span><span class="se">\t</span><span class="s2">percentage_gaps</span><span class="se">\t</span><span class="s2">dbxrefs</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Written </span><span class="si">{}</span><span class="s2"> entries.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)))</span>

<span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;file_samples/PF00870_seed.sth&quot;</span>
<span class="n">outFile</span> <span class="o">=</span> <span class="s2">&quot;file_samples/PF00870_seed.tsv&quot;</span>

<span class="n">infos</span> <span class="o">=</span> <span class="n">getStockholmInfo</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">writeStockholmInfo</span><span class="p">(</span><span class="n">outFile</span><span class="p">,</span> <span class="n">infos</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Alignment with 5 rows and 197 columns
SSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLA...ENL P53_HUMAN/95-289
STVPETSDYPGDHGFRLRFPQSGTAKSVTCTYSPDLNKLFCQLA...SNF P53_DANRE/63-257
-AVPSTEDYAGSYGLKLEFQQNGTAKSVTCTYSTDLNKLFCQLA...DN- F7A9U0_XENTR/70-263
PVIPSNTDYPGPHHFEVTFQQSSTAKSATWTYSPLLKKLYCQIA...DHY P73_HUMAN/113-309
TTVPVTTDYPGSYELELRFQKSGTAKSVTSTYSETLNKLYCQLA...ESR P53_ORYLA/80-270
ID: P53_HUMAN/95-289 start: 95 end: 289 gaps: 2 % gaps: 1.03 refs: 210 PDB; 3Q06 B; 96-289;,PDB; 4MZR B; 95-289;,PDB; 5LAP B; 95-289;
ID: P53_DANRE/63-257 start: 63 end: 257 gaps: 2 % gaps: 1.03 refs: 0
ID: F7A9U0_XENTR/70-263 start: 70 end: 263 gaps: 3 % gaps: 1.55 refs: 0
ID: P73_HUMAN/113-309 start: 113 end: 309 gaps: 0 % gaps: 0.00 refs: 45 PDB; 4A63 E; 114-309;,PDB; 4G83 A; 115-309;,PDB; 3VD2 A; 115-309;
ID: P53_ORYLA/80-270 start: 80 end: 270 gaps: 6 % gaps: 3.16 refs: 0

Written 5 entries.
</pre></div></div>
</div>
</div><ol class="arabic simple" start="3">
<li><p>Given the sequence of the Spike protein in an isolate of the Sars-Cov-2 in <a class="reference external" href="file_samples/Spike_sars_cov2_human.fasta">human</a> and in <a class="reference external" href="file_samples/Spike_sars_cov2_pangolin.fasta">pangolin</a> in fasta, read and store the sequences from the fasta files and align them with <code class="docutils literal notranslate"><span class="pre">pairwise2.align.globalxx</span></code> and check the percentage of identity between the two. Repeat the same analysis with the complete genome of the isolate from Wuhan <a class="reference external" href="file_samples/whuan_sarscov2_human.fasta">wuhan</a> and from
<a class="reference external" href="file_samples/RaTG13_bat_coronavirus_isolate.fasta">bat</a>. <strong>Do not forget to translate them or you might use too much memory for your machine!!!</strong> Note that these are fasta files of the genomic sequence, therefore you should use biopython to translate them into proteins. To solve this exercise implement the following functions:</p></li>
</ol>
<ol class="loweralpha simple">
<li><p><code class="docutils literal notranslate"><span class="pre">readSingleFasta(filename)</span></code>: reads the fasta and returns a SeqRecord;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alignSequences(s1,s2)</span></code> : gets the two sequence records and aligns them returning the alignment;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">getPercentageIdentity(s1,s2):</span></code> gets two aligned sequences and returns the percentage of identity between the two;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">printAlignDetails(alignments,</span> <span class="pre">num_entries</span> <span class="pre">=</span> <span class="pre">1,</span> <span class="pre">ends_size</span> <span class="pre">=</span> <span class="pre">10)</span></code> : that gets the alignments and prints the first <code class="docutils literal notranslate"><span class="pre">num_entries</span></code> plotting the alignment of the initial/final <code class="docutils literal notranslate"><span class="pre">ends_size</span></code> bases, the percentage of identity between the two sequences, the alignment score, the start and end point.</p></li>
</ol>
<div class="tggle" onclick="toggleVisibility(&#39;ex3_cov&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex3_cov" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.pairwise2</span> <span class="kn">import</span> <span class="n">format_alignment</span>



<span class="k">def</span> <span class="nf">readSingleFasta</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">record</span>


<span class="k">def</span> <span class="nf">alignSequences</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">):</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="n">s1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span><span class="n">s2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">alignments</span>

<span class="k">def</span> <span class="nf">getPercentageIdentity</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">):</span>
    <span class="c1">#We assume the length of the two sequences is the same, if not</span>
    <span class="c1">#the following will give an error!</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
    <span class="n">cnt</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s1</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">s1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">s2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
            <span class="n">cnt</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">cnt</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">printAlignDetails</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">num_entries</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ends_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">num_entries</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">):</span>
        <span class="n">num_entries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_entries</span><span class="p">):</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">...</span><span class="si">{}</span><span class="se">\n</span><span class="si">{}</span><span class="s2">...</span><span class="si">{}</span><span class="se">\n</span><span class="si">% i</span><span class="s2">dentity:</span><span class="si">{:.2f}</span><span class="se">\n</span><span class="s2">Score:</span><span class="si">{}</span><span class="se">\n</span><span class="s2">Start: </span><span class="si">{}</span><span class="s2"> End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">seqA</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">ends_size</span><span class="p">],</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">seqA</span><span class="p">[</span><span class="o">-</span><span class="n">ends_size</span><span class="p">:],</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">seqB</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">ends_size</span><span class="p">],</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">seqB</span><span class="p">[</span><span class="o">-</span><span class="n">ends_size</span><span class="p">:],</span>
                                                   <span class="n">getPercentageIdentity</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">seqA</span><span class="p">,</span><span class="n">a</span><span class="o">.</span><span class="n">seqB</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                                                   <span class="n">a</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
             <span class="p">)</span>

<span class="n">S_human</span> <span class="o">=</span> <span class="s2">&quot;file_samples/Spike_sars_cov2_human.fasta&quot;</span>
<span class="n">S_pangolin</span> <span class="o">=</span> <span class="s2">&quot;file_samples/Spike_sars_cov2_pangolin.fasta&quot;</span>

<span class="n">spike_human</span> <span class="o">=</span> <span class="n">readSingleFasta</span><span class="p">(</span><span class="n">S_human</span><span class="p">)</span>
<span class="n">spike_pangolin</span> <span class="o">=</span> <span class="n">readSingleFasta</span><span class="p">(</span><span class="n">S_pangolin</span><span class="p">)</span>

<span class="n">A</span> <span class="o">=</span> <span class="n">alignSequences</span><span class="p">(</span><span class="n">spike_human</span><span class="p">,</span> <span class="n">spike_pangolin</span><span class="p">)</span>
<span class="n">printAlignDetails</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">ends_size</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">num_entries</span><span class="o">=</span> <span class="mi">3</span><span class="p">)</span>


<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">-------------------------------&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;--- Bat vs human Sars-Cov-2 ---&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-------------------------------&quot;</span><span class="p">)</span>
<span class="n">cov_human</span> <span class="o">=</span> <span class="s2">&quot;file_samples/whuan_sarscov2_human.fasta&quot;</span>
<span class="n">cov_bat</span> <span class="o">=</span> <span class="s2">&quot;file_samples/RaTG13_bat_coronavirus_isolate.fasta&quot;</span>

<span class="n">cov_human</span> <span class="o">=</span> <span class="n">readSingleFasta</span><span class="p">(</span><span class="n">cov_human</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;N&quot;</span> <span class="c1">#adding an N to make it divisible by 3</span>
<span class="n">cov_bat</span> <span class="o">=</span> <span class="n">readSingleFasta</span><span class="p">(</span><span class="n">cov_bat</span><span class="p">)</span>


<span class="n">cov_prot_human</span> <span class="o">=</span> <span class="n">cov_human</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="n">cov_prot_bat</span> <span class="o">=</span> <span class="n">cov_bat</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>

<span class="n">A</span> <span class="o">=</span> <span class="n">alignSequences</span><span class="p">(</span><span class="n">cov_prot_human</span><span class="p">,</span> <span class="n">cov_prot_bat</span><span class="p">)</span>
<span class="n">printAlignDetails</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">ends_size</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">num_entries</span><span class="o">=</span> <span class="mi">3</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
M-FVFLVL--LP---LVS-SQCVNLTT-RT...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
MLF-F---FFL-HFALV-NSQCVNL-TGR-...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
% identity:84.18
Score:1160.0
Start: 0 End: 1378
M-FVFLVL-LP---LVS-SQCVNLTT-RT-...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
MLF-F--FFL-HFALV-NSQCVNL-TGR-A...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
% identity:84.24
Score:1160.0
Start: 0 End: 1377
M-FVFLVL-LP---LVS-SQCVNLTT-RT-...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
MLF-F-F-FL-HFALV-NSQCVNL-TGR-A...LKGCCSCGSCCKFDEDDSEPVLKGVKLHYT
% identity:84.24
Score:1160.0
Start: 0 End: 1377


-------------------------------
--- Bat vs human Sars-Cov-2 ---
-------------------------------
IKGLYLP-R*QTNQ-LSISCRSVL*TNFKI...PM*F**LLRRMTKKKKKKKKK-------KX
IKGLYL-SR*QTN-ELSISCRSVL*TNFKI...-------L-------------IAS*ENDK-
% identity:77.94
Score:8725.0
Start: 0 End: 11194
IKGLYLPR*QTNQ-LSISCRSVL*TNFKIC...PM*F**LLRRMTKKKKKKKKK-------KX
IKGLYLSR*QTN-ELSISCRSVL*TNFKIC...-------L-------------IAS*ENDK-
% identity:77.95
Score:8725.0
Start: 0 End: 11193
IKGLYLP-R*QTNQLSISCRSVL*TNFKIC...PM*F**LLRRMTKKKKKKKKK-------KX
IKGLYL-SR*QTNELSISCRSVL*TNFKIC...-------L-------------IAS*ENDK-
% identity:77.95
Score:8725.0
Start: 0 End: 11193
</pre></div></div>
</div>
</div><ol class="arabic simple" start="4">
<li><p>Given a multiple sequence alignment stored in “phylip” format, write three methods: <em>readAlignment</em> that reads the input file and prints the number of sequences present, <em>printAlignments(alignments)</em> that prints the alignments to the screen and <em>computeConsensus(alignments,minFrequency)</em> that creates the consensus of all the alignments. MinFrequency is the minimum frequency (that has to be &gt; 0.5) to keep a base in the consensus. “?” is considered as the consensus if frequency &lt; minFrequency
for all possible bases.</p></li>
</ol>
<p>Ex. if alignments are:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>ATC-G
AAC-G
AAG-G
ATCGT
</pre></div>
</div>
<p>computeConsensus(alignments,0.6) is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>A?C-G
</pre></div>
</div>
<p>Test the script with the file <a class="reference external" href="file_samples/alpha-globin.phy">alpha-globin.phy</a>.</p>
<div class="tggle" onclick="toggleVisibility(&#39;ex3a&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex3a" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">reset</span> -s -f
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">NOTE:</span>
<span class="sd">Three different solutions are provided for computeConsensus.</span>
<span class="sd">One with standard methods, one with set and one with collections.Counter</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="c1">#needed for computeConsensusV3</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>

<span class="k">def</span> <span class="nf">readAlignment</span><span class="p">(</span><span class="n">fn</span><span class="p">):</span>
    <span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> alignments present in </span><span class="se">\&quot;</span><span class="si">{}</span><span class="se">\&quot;\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span><span class="n">fn</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">alignments</span>

<span class="k">def</span> <span class="nf">printAlignments</span><span class="p">(</span><span class="n">aldata</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">aldata</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

<span class="k">def</span> <span class="nf">computeConsensus</span><span class="p">(</span><span class="n">aldata</span><span class="p">,</span> <span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.51</span><span class="p">):</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">minFreq</span> <span class="o">&lt;</span> <span class="mf">0.51</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: minFreq (</span><span class="si">{}</span><span class="s2">) not valid!&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">minFreq</span><span class="p">))</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aldata</span><span class="p">[</span><span class="mi">0</span><span class="p">])):</span>
        <span class="n">chrSeq</span> <span class="o">=</span> <span class="n">aldata</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
        <span class="n">singleChars</span> <span class="o">=</span> <span class="p">[(</span><span class="n">chrSeq</span><span class="p">[</span><span class="n">x</span><span class="p">],</span><span class="n">chrSeq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">[</span><span class="n">x</span><span class="p">]))</span>
                           <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">))</span>
                                   <span class="k">if</span> <span class="n">chrSeq</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">chrSeq</span><span class="p">[</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span> <span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">singleChars</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">consensus</span> <span class="o">+=</span><span class="n">chrSeq</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cons</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">singleChars</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">minFreq</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">)]</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span><span class="n">cons</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span> <span class="s2">&quot;?&quot;</span>
    <span class="k">return</span> <span class="n">consensus</span>

<span class="k">def</span> <span class="nf">computeConsensusV2</span><span class="p">(</span><span class="n">aldata</span><span class="p">,</span> <span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.51</span><span class="p">):</span>
    <span class="c1">#this solution uses sets.</span>
    <span class="c1">#set(list(&quot;ATTTAC&quot;)) returns {&#39;A&#39;, &#39;T&#39;, &#39;C&#39;}</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">minFreq</span> <span class="o">&lt;</span> <span class="mf">0.51</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: minFreq (</span><span class="si">{}</span><span class="s2">) not valid!&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">minFreq</span><span class="p">))</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aldata</span><span class="p">[</span><span class="mi">0</span><span class="p">])):</span>
        <span class="n">chrSeq</span> <span class="o">=</span> <span class="n">aldata</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
        <span class="n">singleChars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">singleChars</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">consensus</span> <span class="o">+=</span><span class="n">chrSeq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cons</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">singleChars</span> <span class="k">if</span> <span class="n">chrSeq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">minFreq</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">)]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span><span class="n">cons</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span> <span class="s2">&quot;?&quot;</span>
    <span class="k">return</span> <span class="n">consensus</span>

<span class="k">def</span> <span class="nf">computeConsensusV3</span><span class="p">(</span><span class="n">aldata</span><span class="p">,</span> <span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.51</span><span class="p">):</span>
    <span class="c1">#this solution uses collections.Counter</span>
    <span class="c1">#collections.Counter(&quot;ATTTAC&quot;) returns the dictionary: {&#39;T&#39;: 3, &#39;A&#39;: 2, &#39;C&#39;: 1}</span>
    <span class="c1">#</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">minFreq</span> <span class="o">&lt;</span> <span class="mf">0.51</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: minFreq (</span><span class="si">{}</span><span class="s2">) not valid!&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">minFreq</span><span class="p">))</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aldata</span><span class="p">[</span><span class="mi">0</span><span class="p">])):</span>
        <span class="n">chrSeq</span> <span class="o">=</span> <span class="n">aldata</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
        <span class="n">singleChars</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">singleChars</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">consensus</span> <span class="o">+=</span><span class="n">chrSeq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cons</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">singleChars</span> <span class="k">if</span> <span class="n">singleChars</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">minFreq</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">chrSeq</span><span class="p">)]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span><span class="n">cons</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">consensus</span> <span class="o">+=</span> <span class="s2">&quot;?&quot;</span>
    <span class="k">return</span> <span class="n">consensus</span>


<span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;file_samples/alpha-globin.phy&quot;</span>
<span class="n">als</span> <span class="o">=</span> <span class="n">readAlignment</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">printAlignments</span><span class="p">(</span><span class="n">als</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">computeConsensus</span><span class="p">(</span><span class="n">als</span><span class="p">,</span><span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Consensus:</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out</span><span class="p">))</span>
<span class="n">out1</span> <span class="o">=</span> <span class="n">computeConsensusV2</span><span class="p">(</span><span class="n">als</span><span class="p">,</span><span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">)</span>
<span class="c1">#print(&quot;Consensus:\n{}&quot;.format(out1))</span>
<span class="n">out2</span> <span class="o">=</span> <span class="n">computeConsensusV3</span><span class="p">(</span><span class="n">als</span><span class="p">,</span><span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">)</span>
<span class="c1">#print(&quot;Consensus:\n{}&quot;.format(out2))</span>
<span class="k">assert</span><span class="p">(</span><span class="n">out1</span> <span class="o">==</span> <span class="n">out2</span><span class="p">)</span>
<span class="k">assert</span><span class="p">(</span><span class="n">out</span> <span class="o">==</span> <span class="n">out1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
4 alignments present in &#34;file_samples/alpha-globin.phy&#34;

ENA|BAA205
ATGAGTCTCTCTGATAAGGACAAGGCTGCTGTGAAAGCCCTATGGGCTAAGATCAGCCCCAAAGCCGATGATATCGGCGCTGAAGCTCTCGGCAGAATGCTGACCGTCTACCCTCAGACCAAGACCTACTTCGCTCACTGGGATGACCTGAGCCCTGGGTCCGGTCCTGTGAAGAAGCATGGCAAGGTTATCATGGGTGCAGTGGCCGATGCCGTTTCAAAAATAGACGACCTTGTGGGAGGTCTGGCCTCCCTGAGCGAACTTCATGCTTCCAAGCTGCGTGTTGACCCGGCCAACTTCAAGATCCTCGCACACAATGTCATCGTGGTCATCGGCATGCTCTTCCCTGGAGACTTCCCCCCAGAGGTTCACATGTCAGTTGACAAGTTTTTCCAGAACTTGGCTCTGGCTCTCTCTGAGAAGTACCGCTAA
ENA|CAA284
ATGTCTCTGACCAGGACTGAGAGGACCATCATCCTGTCCCTGTGGAGCAAGATCTCCACACAGGCAGACGTCATTGGCACCGAGACCCTGGAGAGGCTCTTCTCCTGCTACCCGCAGGCCAAGACCTACTTCCCGCACTTCGAC---CTGCACTCGGGCTCCGCGCAGCTGCGCGCGCACGGCTCCAAGGTGGTGGCCGCCGTGGGCGACGCGGTCAAGAGCATCGACAACGTGACGAGCGCGCTGTCCAAGCTGAGCGAGCTGCACGCCTACGTGCTGCGCGTGGACCCGGTCAACTTCAAGTTCCTGTCCCACTGCCTGCTGGTCACGTTGGCCTCGCACTTCCCCGCCGACTTCACGGCCGACGCGCACGCCGCCTGGGACAAGTTCCTGTCCATCGTGTCCGGCGTCCTGACGGAGAAGTACCGCTGA
ENA|CAA237
ATGGTGCTGTCTCCTGCCGACAAGACCAACGTCAAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCTGGCGAGTATGGTGCGGAGGCCCTGGAGAGGATGTTCCTGTCCTTCCCCACCACCAAGACCTACTTCCCGCACTTCGAC---CTGAGCCACGGCTCTGCCCAAGTTAAGGGCCACGGCAAGAAGGTGGCCGACGCGCTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCACGCGCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGCTCCTAAGCCACTGCCTGCTGGTGACCCTGGCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCTTCCCTGGACAAGTTCCTGGCTTCTGTGAGCACCGTGCTGACCTCCAAATACCGTTAA
ENA|CAA240
ATGGTGCTCTCTGGGGAAGACAAAAGCAACATCAAGGCTGCCTGGGGGAAGATTGGTGGCCATGGTGCTGAATATGGAGCTGAAGCCCTGGAAAGGATGTTTGCTAGCTTCCCCACCACCAAGACCTACTTTCCTCACTTTGAT---GTAAGCCACGGCTCTGCCCAGGTCAAGGGTCACGGCAAGAAGGTCGCCGATGCGCTGGCCAGTGCTGCAGGCCACCTCGATGACCTGCCCGGTGCCTTGTCTGCTCTGAGCGACCTGCATGCCCACAAGCTGCGTGTGGATCCCGTCAACTTCAAGCTCCTGAGCCACTGCCTGCTGGTGACCTTGGCTAGCCACCACCCTGCCGATTTCACCCCCGCGGTACATGCCTCTCTGGACAAATTCCTTGCCTCTGTGAGCACCGTGCTGACCTCCAAGTACCGTTAA
[(&#39;A&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 1), (&#39;T&#39;, 2)]
[(&#39;T&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;T&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;A&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;G&#39;, 3), (&#39;A&#39;, 1)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 3), (&#39;G&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;T&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 2), (&#39;G&#39;, 1)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 3), (&#39;G&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 1), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;G&#39;, 1)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 1), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 2), (&#39;G&#39;, 1), (&#39;C&#39;, 1)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 1), (&#39;A&#39;, 1)]
[(&#39;A&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;T&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 1), (&#39;A&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 1), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 2), (&#39;A&#39;, 1)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 2), (&#39;G&#39;, 1), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 2), (&#39;A&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;A&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;A&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;-&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;-&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;-&#39;, 3)]
[(&#39;C&#39;, 3), (&#39;G&#39;, 1)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 3), (&#39;A&#39;, 1)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;T&#39;, 1), (&#39;C&#39;, 1)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 1), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 2), (&#39;A&#39;, 2)]
[(&#39;A&#39;, 3), (&#39;G&#39;, 1)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;T&#39;, 3), (&#39;C&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 1), (&#39;A&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 2), (&#39;G&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 2), (&#39;C&#39;, 1)]
[(&#39;A&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;A&#39;, 2), (&#39;G&#39;, 1), (&#39;C&#39;, 1)]
[(&#39;T&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 2), (&#39;C&#39;, 1)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 1), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 3), (&#39;C&#39;, 1)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 1), (&#39;T&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 2), (&#39;G&#39;, 1)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;T&#39;, 3), (&#39;A&#39;, 1)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;G&#39;, 1), (&#39;T&#39;, 1)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 2), (&#39;A&#39;, 1)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 3), (&#39;T&#39;, 1)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;G&#39;, 3), (&#39;A&#39;, 1)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 2), (&#39;T&#39;, 1)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;A&#39;, 1), (&#39;T&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;T&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;C&#39;, 2), (&#39;G&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 1), (&#39;A&#39;, 2)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 1), (&#39;C&#39;, 3)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;T&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;C&#39;, 1), (&#39;G&#39;, 2)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 4)]
[(&#39;C&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;C&#39;, 4)]
[(&#39;T&#39;, 1), (&#39;G&#39;, 1), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;A&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;G&#39;, 2), (&#39;C&#39;, 2)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;A&#39;, 1), (&#39;G&#39;, 3)]
[(&#39;T&#39;, 4)]
[(&#39;A&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;C&#39;, 4)]
[(&#39;G&#39;, 4)]
[(&#39;C&#39;, 2), (&#39;T&#39;, 2)]
[(&#39;T&#39;, 4)]
[(&#39;G&#39;, 1), (&#39;A&#39;, 3)]
[(&#39;A&#39;, 4)]

Consensus:
ATG???CT?TCT??????GACAAGACCA?C?TCAAGGCC???TGGGG?AAGATC?GC?C?CA?GC?G??GA???TGG?GC?GA?GCCCTGGA?AGGATGTT??C???CT?CCC????ACCAAGACCTACTTCCC?CACTT?GA?---CTGAGCC??GGCTC?GC?CA?GT?AAGG??CACGGCAAGAAGGT?G??G??GC??TGGCC?A?GC?GT?????AC?T?GACGAC?TG?C??G?GC?CTGTCC?C?CTGAGCGA?CTGCA?GC??ACAAGCTGCG?GTGGACCCGGTCAACTTCAAG?TCCT???CCACTGCCTGCTGGTGACC?TGGCC???CAC?TCCC?GCCGA?TTCACCCC?G?GGT?CACGC?TC??TGGACAAGTTCCT??C????GTG??C??CGT?CTGAC????AAGTACCG?TAA
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Reminder of the new things seen...</span>

<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>

<span class="n">my_seq</span> <span class="o">=</span> <span class="s2">&quot;ATTAGATCACATAAAA&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">my_seq</span><span class="p">))</span>

<span class="c1">#The set data structure</span>
<span class="n">S</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The set: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">S</span><span class="p">))</span>

<span class="c1">#The counter object</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The counts:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counts</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">minFreq</span> <span class="o">=</span> <span class="mf">0.51</span>
<span class="nb">print</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">counts</span> <span class="k">if</span> <span class="n">counts</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">minFreq</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)])</span>

<span class="c1">#Asserts to make sure some conditions hold...</span>
<span class="k">assert</span><span class="p">(</span><span class="mi">10</span> <span class="o">&gt;</span> <span class="mi">5</span><span class="p">)</span>
<span class="k">assert</span><span class="p">(</span><span class="n">my_seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="k">assert</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">())))</span>
<span class="k">assert</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">())))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Sequence: ATTAGATCACATAAAA
The set: {&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;}
The counts:Counter({&#39;A&#39;: 9, &#39;T&#39;: 4, &#39;C&#39;: 2, &#39;G&#39;: 1})

[&#39;A&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">AssertionError</span>                            Traceback (most recent call last)
<span class="ansi-green-fg">&lt;ipython-input-31-2e6861380230&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">     21</span> <span class="ansi-green-fg">assert</span><span class="ansi-blue-fg">(</span>my_seq<span class="ansi-blue-fg">[</span><span class="ansi-cyan-fg">0</span><span class="ansi-blue-fg">]</span> <span class="ansi-blue-fg">==</span> <span class="ansi-blue-fg">&#39;A&#39;</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     22</span> <span class="ansi-green-fg">assert</span><span class="ansi-blue-fg">(</span>len<span class="ansi-blue-fg">(</span>my_seq<span class="ansi-blue-fg">)</span> <span class="ansi-blue-fg">==</span> sum<span class="ansi-blue-fg">(</span>list<span class="ansi-blue-fg">(</span>counts<span class="ansi-blue-fg">.</span>values<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-fg">---&gt; 23</span><span class="ansi-red-fg"> </span><span class="ansi-green-fg">assert</span><span class="ansi-blue-fg">(</span>len<span class="ansi-blue-fg">(</span>my_seq<span class="ansi-blue-fg">)</span> <span class="ansi-blue-fg">&gt;</span> sum<span class="ansi-blue-fg">(</span>list<span class="ansi-blue-fg">(</span>counts<span class="ansi-blue-fg">.</span>values<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>

<span class="ansi-red-fg">AssertionError</span>:
</pre></div></div>
</div>
</div><ol class="arabic simple" start="5">
<li><p>Load the contigs present in the <a class="reference external" href="file_samples/filtered_contigs.fasta">filtered_contigs.fasta</a> file and translate each DNA sequence into the corresponding protein. Count the number of stop codons (i.e. <a href="#id1"><span class="problematic" id="id2">*</span></a>) for each sequence and print them to the user (e.g. MDC020656.85 51). Finally, write the translated proteins in another .fasta file (e.g. filtered_contigs_translated.fasta).</p></li>
</ol>
<div class="tggle" onclick="toggleVisibility(&#39;ex3&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex3" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">reset</span> -s -f

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">seqIterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/filtered_contigs.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="n">sRcds</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqIterator</span><span class="p">:</span>
    <span class="c1">#to avoid problems if not divisible by 3!</span>
    <span class="n">seq1</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">tomutable</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">seq1</span> <span class="o">+=</span><span class="s2">&quot;N&quot;</span>
    <span class="n">transl</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">toseq</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
    <span class="n">stop_cnt</span> <span class="o">=</span> <span class="n">transl</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">stop_cnt</span><span class="p">)</span>
    <span class="n">s</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">transl</span>
    <span class="n">sRcds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">N</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sRcds</span><span class="p">,</span><span class="s2">&quot;file_samples/filtered_contigs_translated.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> sequences written to output file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MDC001115.177    65
MDC013284.379    93
MDC018185.243    418
MDC018185.241    467
MDC004527.213    48
MDC012176.157    30
MDC001204.810    155
MDC004389.256    43
MDC018297.229    317
MDC001802.364    45
MDC014057.243    23
MDC021015.302    19
MDC017187.314    83
MDC012865.410    68
MDC000427.83     42
MDC017187.319    251
MDC004364.265    145
MDC002360.219    60
MDC015155.172    47
MDC019140.398    67
MDC019140.399    56
MDC011390.337    71
MDC007154.375    248
MDC010588.505    29
MDC002519.240    500
MDC006346.711    376
MDC011551.182    806
MDC002717.156    55
MDC006346.719    157
MDC007838.447    141
MDC007018.186    47
MDC017873.233    35
MDC016296.138    176
MDC019067.226    49
MDC014019.318    45
MDC007995.528    53
MDC026961.60     71
MDC020963.161    65
MDC005174.220    132
MDC040033.7      142
MDC019674.147    89
MDC010450.877    40
MDC007097.457    89
MDC016278.70     69

44 sequences written to output file
</pre></div></div>
</div>
</div><ol class="arabic simple" start="6">
<li><p>Write the following three functions:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">countMatches(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements match;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countMismatches(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements are different (i.e. they are not gaps but the character is different);</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countGapOpens(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap opened in the alignment;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countGapExtensions(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap extensions in the alignment</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">getScore(s1,s2,</span> <span class="pre">matchScore,</span> <span class="pre">mismatchPenalty,</span> <span class="pre">gapOpenPenalty,</span> <span class="pre">gapExtensionPenalty)</span></code> and returns the score of the alignment of s1 and s2 (aligned with pairwise2 as before) given the scores and penalties in input.</p></li>
</ol>
</li>
</ol>
<p>Finally, align the sequences of the <a class="reference external" href="https://en.wikipedia.org/wiki/Interleukin_12">Interleukin-12</a> chain A (let’s call it s1) <a class="reference external" href="file_samples/IL12A.fasta">IL12A.fasta</a> and B <a class="reference external" href="file_samples/IL12B.fasta">IL12B.fasta</a> (let’s call it s2) with the following parameters:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>pairwise2.align.globalxx(s1,s2)
pairwise2.align.globalms(s1,s2,,1,-0.1,-1,-0.1)
pairwise2.align.globalxs(s1, s2,1,-0.5,-5,-0.5)
pairwise2.align.globalxd(s1, s2,1,-0.5,-5,-0.5,-5,-5)
</pre></div>
</div>
<p>and check the score as computed from pairwise2 and from your functions (use the first of all the alignments to check). Note the effect of changing parameters.</p>
<div class="tggle" onclick="toggleVisibility(&#39;ex4&#39;);"><p>Show/Hide Solution</p>
</div><div id="ex4" style="display:none;"><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">reset</span> -s -f

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>


<span class="k">def</span> <span class="nf">countMatches</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a1</span><span class="p">))</span> <span class="k">if</span> <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>

<span class="k">def</span> <span class="nf">countMismatches</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a1</span><span class="p">))</span> <span class="k">if</span> <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">and</span> <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">])</span>

<span class="k">def</span> <span class="nf">countGapOpens</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">):</span>
    <span class="n">gaps</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">a1</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="c1">#if ((a1[i] == &quot;-&quot; or a2[i] == &quot;-&quot;) and a1[i-1] != &quot;-&quot; and a2[i-1] !=&quot;-&quot;)</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span><span class="s2">&quot;-&quot;</span><span class="p">))</span>
           <span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gaps</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">countGapExtensions</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">):</span>
    <span class="n">gaps</span> <span class="o">=</span> <span class="p">[</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">a1</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span><span class="s2">&quot;-&quot;</span><span class="p">))</span>
           <span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gaps</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">getScore</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">,</span> <span class="n">matchScore</span><span class="p">,</span> <span class="n">mismatchPenalty</span><span class="p">,</span> <span class="n">gapOpenPenalty</span><span class="p">,</span> <span class="n">gapExtensionPenalty</span><span class="p">):</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">countMatches</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
    <span class="n">mis</span> <span class="o">=</span> <span class="n">countMismatches</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
    <span class="n">gapo</span> <span class="o">=</span> <span class="n">countGapOpens</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
    <span class="n">gape</span> <span class="o">=</span> <span class="n">countGapExtensions</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
    <span class="n">score</span> <span class="o">=</span> <span class="n">matchScore</span><span class="o">*</span><span class="n">mat</span> <span class="o">+</span> <span class="n">mis</span><span class="o">*</span><span class="n">mismatchPenalty</span> <span class="o">+</span> <span class="n">gapo</span><span class="o">*</span><span class="n">gapOpenPenalty</span> <span class="o">+</span> <span class="n">gape</span><span class="o">*</span><span class="n">gapExtensionPenalty</span>
    <span class="k">return</span> <span class="n">score</span>

<span class="k">def</span> <span class="nf">getSequenceData</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span>

<span class="k">def</span> <span class="nf">printInfo</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">computedScore</span><span class="p">,</span><span class="n">Mscore</span><span class="p">,</span><span class="n">mScore</span><span class="p">,</span><span class="n">goScore</span><span class="p">,</span><span class="n">geScore</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a2</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">computedScore</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">M</span> <span class="o">=</span> <span class="n">countMatches</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">countMismatches</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
    <span class="n">go</span> <span class="o">=</span> <span class="n">countGapOpens</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
    <span class="n">ge</span> <span class="o">=</span> <span class="n">countGapExtensions</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Matches:</span><span class="si">{}</span><span class="s2"> Mismatches:</span><span class="si">{}</span><span class="s2"> GapOpens:</span><span class="si">{}</span><span class="s2"> GapExtensions:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">M</span><span class="p">,</span><span class="n">m</span><span class="p">,</span><span class="n">go</span><span class="p">,</span><span class="n">ge</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Recomputed Score: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">getScore</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">Mscore</span><span class="p">,</span><span class="n">mScore</span><span class="p">,</span><span class="n">goScore</span><span class="p">,</span><span class="n">geScore</span><span class="p">)))</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">getSequenceData</span><span class="p">(</span><span class="s2">&quot;file_samples/IL12A.fasta&quot;</span><span class="p">)</span>
<span class="n">s2</span> <span class="o">=</span> <span class="n">getSequenceData</span><span class="p">(</span><span class="s2">&quot;file_samples/IL12B.fasta&quot;</span><span class="p">)</span>

<span class="n">aligns</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
<span class="n">a1</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">a2</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sc</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
<span class="n">printInfo</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">sc</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>


<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">aligns</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalms</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">a1</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">a2</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sc</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
<span class="n">printInfo</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">sc</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">)</span>


<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">aligns</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalms</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>

<span class="n">a1</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">a2</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sc</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
<span class="n">printInfo</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">sc</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>

<span class="n">aligns</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalmd</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>

<span class="n">a1</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">a2</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sc</span> <span class="o">=</span> <span class="n">aligns</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
<span class="n">printInfo</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">sc</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MCPAR------S---L--L---LVAT---L---VL----LDHLSL----ARNLP---VA--TP-D-P---GMFPC----LHHS-QNLLRAVSNM---LQ---KARQTL-----EF-----YPCTSE-----EID---H--------ED-I--T---KD-KTSTVEACLP-----L--ELT-KNE-S----C--LNSRETSF-I-TN----------GS-------CL-A---S--R-----KT----SFMMAL--CL---S-------S--IYE---D----LKMYQVE-----F------KTMNA----K-L-LMD-P-K--RQIFLDQNMLAVIDELMQALNFN-S-E---TV---PQK-S--SLE------------EP--D--FYKT-K-----IKLCILLHAFR----I--RAVTI-DRVMSYLN--------AS----
MC---HQQLVISWFSLVFLASPLVA-IWELKKDV-YVVELD----WYPDA---PGEMV-VLT-CDTPEEDG----ITWTL---DQ------S--SEVL-GSGK---TLTIQVKEFGDAGQY--T--CHKGGE--VLSHSLLLLHKKEDGIWSTDILKDQK----E---PKNKTFLRCE--AKN-YSGRFTCWWL----T--TIST-DLTFSVKSSRGSSDPQGVTC-GAATLSAERVRGDNK-EYEYS-----VEC-QEDSACPAAEESLPI-EVMVDAVHKLK-Y--ENYTSSFFIRDIIK----PDPPKNLQL--KPLKNSR-----Q----V--E--------VSWEYPDT-WSTP--HSYFSL-TFCVQVQGKSKRE-KKDRVF--TDKTSATVI--C------RKNASISVRA---QDR---Y--YSSSWSEWASVPCS
103.0

Matches:103 Mismatches:0 GapOpens:117 GapExtensions:223
Recomputed Score: 103

MC------PARSL--L---LVA--TL------VLLDHLSLARNLPVA-TPD-P---GMFPCLHHSQNLLRAVSNM-LQKARQTLEF-----YPC------TS---------EE-IDHEDITKD-KTSTVEACLPLELTKNES----CLNSRETSFITNGSCLASRKTS-----FMMALCLS--SIYEDLK--MYQVEFKTMNA------------------KL---------------LMDP-KR-QIFLDQNMLAV------IDELMQALNFNSET----VPQKSSLEEPD--FY-KTKIKLCILLHAFRI--RAVTIDRVMS---YLNA----S
MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVL--GSGKTL--TIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCE-AKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKT-SATVICRKNASISVRA--QDRYYSSSWSEWASVPCS
28.19999999999986

Matches:84 Mismatches:127 GapOpens:34 GapExtensions:91
Recomputed Score: 28.199999999999996

MCPARSLLLVATLVLLDHLSLA----------RNLPVATPDPGMFPCLHHSQNLLRAVSNMLQKARQTL----EFYPCTSEEIDHEDITKDKTSTVEACLPLELTKNESCLNS-------RETSFITNGSCLASRKTSFMMALCLSSIYEDL----------------------------------KMYQVEFKTMNA------------------KLLMDPKRQIFLDQNMLAVIDELMQALNFNSETVPQKSSLEEPDFYKT------------------------------------KIKLCILLHAFRIRAVTIDRVMSYLNAS
MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVLGSGKTLTIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCEAKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKTSATVICRKNASISVRAQDRYYSSSWSEWASVPCS
-122.0

Matches:46 Mismatches:173 GapOpens:6 GapExtensions:103
Recomputed Score: -122.0
MCPARSLLLVATLVLLDHLSLA----------RNLPVATPDPGMFPCLHHSQNLLRAVSNMLQKARQTL----EFYPCTSEEIDHEDITKDKTSTVEACLPLELTKNESCLNS-------RETSFITNGSCLASRKTSFMMALCLSSIYEDL----------------------------------KMYQVEFKTMNA------------------KLLMDPKRQIFLDQNMLAVIDELMQALNFNSETVPQKSSLEEPDFYKT------------------------------------KIKLCILLHAFRIRAVTIDRVMSYLNAS
MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVLGSGKTLTIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCEAKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKTSATVICRKNASISVRAQDRYYSSSWSEWASVPCS
-122.0

Matches:46 Mismatches:173 GapOpens:6 GapExtensions:103
Recomputed Score: -122.0
</pre></div></div>
</div>
</div></div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="M1_practical10b.html" class="btn btn-neutral float-right" title="Module 1, Practical 10b" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="M1_practical9.html" class="btn btn-neutral float-left" title="Module 1, Practical 9" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright # 2021 - 2021, Luca Marchetti

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>