{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Module 1, midterm simulation test\n",
    "\n",
    "## Available material for exams\n",
    "\n",
    "The course notes can be used as the slides of the theory and practicals. Documentation on the libraries are here:\n",
    "\n",
    "* [matplotlib](material/Matplotlib.pdf)\n",
    "* [numpy](material/numpy-ref.pdf)\n",
    "* [pandas](material/pandas.pdf)\n",
    "* [biopython](material/biopython.pdf)\n",
    "\n",
    "You can download all of them here: [archive](material/material.zip)\n",
    "\n",
    "## Download the data\n",
    "\n",
    "1) Download  [sciprog-qcb-2021-12-05-FIRSTNAME-LASTNAME-ID.zip](MTsim/sciprog-qcb-2021-12-05-FIRSTNAME-LASTNAME-ID.zip) and extract it *on your desktop*. Folder content should be like this:\n",
    "\n",
    "```\n",
    "|- sciprog-qcb-2021-12-05-FIRSTNAME-LASTNAME-ID\n",
    "    |- QCB_midTerm_simulation.pdf\n",
    "    |- exercise1.py\n",
    "    |- exercise2.py\n",
    "    |- data_reduced.agp\n",
    "    |- small_seq.fasta\n",
    "    |- small.agp\n",
    "```\n",
    "\n",
    "2) Rename `sciprog-qcb-2021-12-05-FIRSTNAME-LASTNAME-ID` folder: put your name, lastname an id number, like `sciprog-qcb-2021-12-05-john-doe-432432`\n",
    "\n",
    "From now on, you will be editing the files in that folder. At the end of the exam, that is what will be evaluated.\n",
    "\n",
    "3) Edit the files following the instructions.\n",
    "\n",
    "## Problem 1\n",
    "\n",
    "Given a list of positive integers (possibly repeated and unsorted) in the range [1,N], write a function that finds the missing values and returns them as a list. **Note that the function should not crash if the list is empty**. A warning should also be printed in case the user by mistake had negative numbers in the list.\n",
    "\n",
    "Ex. \n",
    "```\n",
    "S = [1,9,7, 7, 4, 3, 3, 3]\n",
    "S1 = list(range(10))\n",
    "print(find_missing(S))\n",
    "print(find_missing(S1))\n",
    "print(find_missing([]))\n",
    "S2 = [1, -72, 4, -3, -3, 3,10]\n",
    "M = find_missing(S2)\n",
    "print(M)\n",
    "```\n",
    "should return:\n",
    "\n",
    "```\n",
    "[2, 5, 6, 8]\n",
    "[]\n",
    "Warning: list is empty. Returning None\n",
    "None\n",
    "Warning -72 is <0. Ignoring it.\n",
    "Warning -3 is <0. Ignoring it.\n",
    "Warning -3 is <0. Ignoring it.\n",
    "[2, 5, 6, 7, 8, 9]\n",
    "\n",
    "```\n",
    "\n",
    "## Problem 2\n",
    "\n",
    "The .agp file ```data_reduced.agp``` is a compact representation on how a set of assembled contigs made it into the scaffolds. The first few lines are reported below:\n",
    "```\n",
    "ScaffID\ts_start\ts_end\ttype\tcontig\tc_start\tc_end\tc_strand\n",
    "scaffold_1      1       120953  W       scf7180000021845        1       120953  -\n",
    "scaffold_1      120954  121453  N       500     scaffold        yes     na\n",
    "scaffold_1      121454  1026498 W       scf7180000018491_2      1       905045  +\n",
    "scaffold_1      1026499 1026998 N       500     scaffold        yes     na\n",
    "\n",
    "```\n",
    "\n",
    "In particular, the first row states that ```scaffold_1``` from position ```1``` to ```120953``` has been built using the sequence of the contig ```scf7180000021845``` from position ```1``` to ```120953``` in reverse strand (```-```) which means that the sequence has to be **reverse-complemented**. The second row states that in ```scaffold_1``` positions ```120954``` to ```121453``` are a **gap** made of 500 ```N``` (note the 4th column is ```N``` rather ```W``` that stands for whole genome sequence). \n",
    "\n",
    "Let's suppose to have three sequences $s1=\"ATAATA\"$, $s2=\"AAA\"$ and $s3=\"CCAAA\"$, the following agp-formatted entries can be used to create a sequence ```my_scaff```:\n",
    "\n",
    "```\n",
    "my_scaff\t1\t6\tW\ts1\t1\t6\t+\n",
    "my_scaff\t7\t9\tN\t3\tscaffold\tyes\tna\n",
    "my_scaff\t10 \t12\tW\ts2\t1\t3\t-\n",
    "my_scaff\t13\t15\tN\t3\tscaffold\tyes\tna\n",
    "my_scaff\t16\t17\tW\ts3\t1\t2\t+\n",
    "```\n",
    "\n",
    "this would represent a fasta-formatted sequence:\n",
    "```\n",
    ">my_scaff\n",
    "ATAATANNNTTTNNNCC\n",
    "```\n",
    "\n",
    "where basically the sequence is composed by s1 as it is, followed by three N, followed by the reverse complement of s2, three N and the first two characters of s3.\n",
    "\n",
    "\n",
    "The file ```small_seq.fasta``` stores sequence information in .fasta format. A mock entry is the following:\n",
    "```\n",
    ">Chr01 \n",
    "AGGCCTAGGTCTTCCAGAGTCGCTTTTTCCAGCTCCAGACCGATCTCTTCAG\n",
    "AGGCCAATCGCCAGTTTACCACATACACCCAGACCGATCTCTTCAG\n",
    "```\n",
    "where the first line is the identifier of the read and starts with a \">\". The sequence follows the line with the identifier and can be on multiple lines.\n",
    "\n",
    "Implement the following python functions:\n",
    "\n",
    "\n",
    "1. ```computeStats(filename, show_output = True)```: gets the ```filename``` of a .agp file as explained above, stores its content in a suitable data structure of your choice (hint: pandas might help here). If ```show_output``` is ```False``` the function only returns the data structure. Otherwise, it counts (and prints) the total number of entries, the total number of scaffolds (**hint: you can use DataFrame[column].unique()**), total number of contigs (and their total size **note that you might have to convert the c_start and c_end column to int with .astype(int)**) and total number of gaps (and their total size). The function should also produce a box plot of the number of contigs per scaffold.\n",
    "\n",
    "**Note: The function should return the data structure containing all the data.**\n",
    "\n",
    "Calling:\n",
    "```\n",
    "fn = \"data_reduced.agp\"\n",
    "scaffDF = computeStats(fn)\n",
    "```\n",
    "should give:\n",
    "```\n",
    "The file contains 7898 entries\n",
    "... 1958 scaffolds\n",
    "... 4928 contigs (tot. size: 873,456,804 bases)\n",
    "... and 2970 gaps (tot. size: 1,485,000 bases)\n",
    "\n",
    "```\n",
    "![mtplot](MTsim/MTplot.png)\n",
    "\n",
    "2. ```printSequence(scaffInfo, scafID, sequenceFile)```: gets the ```scaffInfo``` data structure **created by computeStats**, a **scaffold identifier** ```scaffID``` and **the filename of a fasta formatted file** ```sequenceFile``` and if ```scafID``` is present in ```scaffInfo``` it **prints a fasta-formatted string reporting the sequence of the scaffold** built as discussed above. \n",
    "\n",
    "**Hint: you can use biophtyon to read the fasta file.** \n",
    "\n",
    "\n",
    "Calling:\n",
    "```\n",
    "scaffDF = computeStats(\"small.agp\", show_output = False)\n",
    "printSequence(scaffDF,\"my_scaff\",\"small_seq.fasta\")\n",
    "print(\"\")\n",
    "printSequence(scaffDF,\"my_scaff2\",\"small_seq.fasta\")\n",
    "print(\"\")\n",
    "printSequence(scaffDF,\"my_other_scaff\",\"small_seq.fasta\")\n",
    "print(\"\")\n",
    "printSequence(scaffDF,\"scaffold3\",\"small_seq.fasta\")\n",
    "```\n",
    "should give:\n",
    "```\n",
    ">my_scaff\n",
    "ATAATANNNTTTNNNCC\n",
    "\n",
    ">my_scaff2\n",
    "TATTTTTATATGTATGTAATNNNNNNNNNNTTTATATATA\n",
    "\n",
    "Warning: scaffold my_other_scaff not present\n",
    "\n",
    ">scaffold3\n",
    "NNNNNNNNNNNNNNNNNNNNCCCCGGAGGTACCTCCGGGGCCCCGGAGGT\n",
    "```\n"
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}