{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Module 1, QCB midterm test, November 12th 2021\n",
    "\n",
    "## Download the data\n",
    "\n",
    "1) Download  [sciprog-qcb-2021-11-12-FIRSTNAME-LASTNAME-ID.zip](MidTerm/sciprog-qcb-2021-11-12-FIRSTNAME-LASTNAME-ID.zip) and extract it *on your desktop*. Folder content should be like this:\n",
    "\n",
    "```\n",
    "|- sciprog-qcb-2021-11-12-FIRSTNAME-LASTNAME-ID\n",
    "    |- QCB_midTerm_2021_11_12.pdf\n",
    "    |- exercise1.py\n",
    "    |- exercise2.py\n",
    "    |- shortSample.fasta\n",
    "    |- test_reads_10k.fasta\n",
    "```\n",
    "\n",
    "2) Rename `sciprog-qcb-2021-11-12-FIRSTNAME-LASTNAME-ID` folder: put your name, lastname an id number, like `sciprog-qcb-2021-11-12-john-doe-432432`\n",
    "\n",
    "From now on, you will be editing the files in that folder. At the end of the exam, that is what will be evaluated.\n",
    "\n",
    "3) Edit the files following the instructions.\n",
    "\n",
    "## Problem 1\n",
    "Given a text string S (possibly on several lines), write a python function that takes the string S and a second string S1 (assumed to be shorter than S) in input and returns a list with all positions at which S1 starts in S. If no such positions exist a warning message should be printed and an empty list returned.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "```\n",
    "S = \"\"\"TGAATACATATTATTCCGC\n",
    "GTCCTACCAAAAAATATACATAGATAT\n",
    "GATACTAGGGGACCGGTTCCGGGATGA\n",
    "TGATGATACTAGGGGACCGGTTGATAC\"\"\"\n",
    "\n",
    "print(findPositions(S, \"TGA\"))\n",
    "print(findPositions(S, \"GAT\"))\n",
    "print(findPositions(S, \"none\"))\n",
    "```\n",
    "\n",
    "should return:\n",
    "\n",
    "```\n",
    "String 'TGA' found at the following positions:\n",
    "[0, 45, 70, 73, 76, 94]\n",
    "\n",
    "String 'GAT' found at the following positions:\n",
    "[41, 46, 68, 71, 74, 77, 95]\n",
    "\n",
    "Warning: string 'none' not found. Returning an empty list.\n",
    "[]\n",
    "```\n",
    "\n",
    "## Problem 2\n",
    "\n",
    "Fasta is a format for storing sequence information and the corresponding sequencing quality. A sample entry is the following:\n",
    "\n",
    "```\n",
    ">SRR190875.100000 HWI-ST180_0167:2:1:6207:136494 \n",
    "AGGCCTAGGTCTTCCAGAGTCGCTTTTTCCAGCTCCAGACCGATCTCTTCAG\n",
    "```\n",
    "\n",
    "where the first line is the identifier of the read and starts with a `>`. The sequence follows the line with the identifier **(in general it can be on several lines, but you can assume it is only on one line).**\n",
    "\n",
    "**HINT: Biopython's `SeqIO.parse` can read fasta formatted files.**\n",
    "\n",
    "Kmers are substrings of length K of a given sequence. For example, 2mers are all substrings having length 2 of a given sequence. Assuming a sequence `seq = \"AATAACTAGC\"`, all possible 2mers of seq are: `AA,AT,TA,AC,CT,AG,GC`. Note that the 2mers `\"AA\"` and `\"TA\"` are present twice.\n",
    "\n",
    "Start editing `exercise2.py`, implementing the required functions as in the next points.\n",
    "\n",
    "## 2.1) `computeKmers(seq, kmerLen, kmerDict)`\n",
    "\n",
    "Implement `computeKmers(seq, kmerLen, kmerDict)` : gets a sequence `seq` (stored as a string), an integer `kmerLen` specifying the size of the kmer and fills the dictionary `kmerDict` with all the possible kmers of length kmerLen present in the sequence as keys and their multiplicity (i.e., the number of times they appear in the sequence) as values.\n",
    "\n",
    "**Example:** given the sequence `seq=\"AATAACTAGC\"` the dictionary of 2mers is the following: \n",
    "\n",
    "```python\n",
    "{'AA': 2, 'AG': 1, 'TA': 2, 'AT': 1, 'GC': 1, 'CT': 1, 'AC': 1}\n",
    "```\n",
    "\n",
    "Note: if `kmerLen` is higher than `len(seq)` the dictionary will be not updated. Calling\n",
    "\n",
    "```python\n",
    "seq = \"AATTAATTAACTAGCCTTAA\"\n",
    "twoMers = dict()\n",
    "print(\"sequence:\" , seq)\n",
    "computeKmers(seq, 2, twoMers)\n",
    "print(\"twoMers\")\n",
    "print(twoMers)\n",
    "fourMers = dict()\n",
    "computeKmers(seq, 4, fourMers)\n",
    "print(\"4mers\")\n",
    "print(fourMers)\n",
    "t2Mers = dict()\n",
    "computeKmers(seq, 32, t2Mers)\n",
    "print(\"32mers\")\n",
    "print(t2Mers)\n",
    "```\n",
    "\n",
    "should give:\n",
    "\n",
    "```\n",
    "sequence: AATTAATTAACTAGCCTTAA\n",
    "twoMers\n",
    "{'AA': 4, 'AT': 2, 'TT': 3, 'TA': 4, 'AC': 1, 'CT': 2, 'AG': 1, 'GC': 1, 'CC': 1}\n",
    "4mers\n",
    "{'AATT': 2, 'ATTA': 2, 'TTAA': 3, 'TAAT': 1, 'TAAC': 1, 'AACT': 1, 'ACTA': 1, 'CTAG': 1, 'TAGC': 1, 'AGCC': 1, 'GCCT': 1, 'CCTT': 1, 'CTTA': 1}\n",
    "32mers\n",
    "{}\n",
    "```\n",
    "\n",
    "## 2.2)  `parseFasta(fileName,kmerLen)`\n",
    "\n",
    "Implement `parseFasta(fileName,kmerLen)`: reads the fasta formatted file `fileName` and returns a dictionary with all the kmers having length `kmerLen` and their multiplicities present in the file. The function should print the number of sequences read from the file and the number of distinct kmers.\n",
    "\n",
    "**Important:** it could be very useful to take advantage of the function `computeKmers` developed at point 1 to implement this function.\n",
    "\n",
    "**Example:** if the file fileName contains the two sequences `\"ATATCACATCTG\"` and `\"CTGACATATAT\"` (with the respective sequence headers), the output dictionary of 4mers would be:\n",
    "\n",
    "```python\n",
    "{'ATAT': 3, 'TCTG': 1, 'TCAC': 1, 'TGAC': 1, 'CATA': 1, 'GACA': 1, 'TATA': 1, 'TATC': 1, 'CATC': 1, 'ATCT': 1, 'CACA': 1, 'ATCA': 1, 'CTGA': 1, 'ACAT': 2}\n",
    "```\n",
    "Given the input files ```shortSample.fasta``` and ```test_reads_10k.fasta``` (that are included in your working folder), calling:\n",
    "\n",
    "```python\n",
    "myFile = \"shortSample.fasta\"\n",
    "kmers = parseFasta(myFile,4)\n",
    "print(kmers)\n",
    "```\n",
    "\n",
    "should give:\n",
    "\n",
    "```\n",
    "Read 2 sequences.\n",
    "Total number of 4mers: 14\n",
    "{'ATAT': 3, 'TATC': 1, 'ATCA': 1, 'TCAC': 1, 'CACA': 1, 'ACAT': 2, 'CATC': 1, 'ATCT': 1, 'TCTG': 1, 'CTGA': 1, 'TGAC': 1, 'GACA': 1, 'CATA': 1, 'TATA': 1}\n",
    "```\n",
    "\n",
    "while\n",
    "\n",
    "```python\n",
    "myFile = \"test_reads_10k.fasta\"\n",
    "kmers = parseFasta(myFile,17)\n",
    "```\n",
    "\n",
    "should print:\n",
    "\n",
    "```\n",
    "Read 10000 sequences.\n",
    "Total number of 17mers: 765263\n",
    "```\n",
    "\n",
    "**IMPORTANT:** given the length of `test_reads_10k.fasta`, try to process this file only when the function `parseFasta()` seems to work with the file `shortSample.fasta`.\n",
    "\n",
    "## 2.3)  `plotKmerSpectrum(kmerDict)`\n",
    "\n",
    "Implement `plotKmerSpectrum(kmerDict)`: given a dictionary `kmerDict` created as above (i.e., keys are kmer sequences and values are the number of occurrences of the kmer sequence) plots the kmer spectrum of all the kmers present in the dictionary. The kmer spectrum is a histogram of the multiplicities of all kmers. For example, given the dictionary\n",
    "\n",
    "```python\n",
    "{'AC': 1, 'AG': 1, 'AA': 4, 'AT': 2, 'CT': 2, 'CC': 1, 'TA': 4, 'TT': 3, 'GC': 1}\n",
    "```\n",
    "\n",
    "a dictionary representing the kmer spectrum of these kmers is:\n",
    "\n",
    "```python\n",
    "{1: 4, 2: 2, 3: 1, 4: 2}\n",
    "```\n",
    "\n",
    "which basically means that there are 4 2mers with multiplicity 1, 2 with multiplicity 2, 1 with multiplicity 3 and 2 with multiplicity 4. The plot for such a kmer spectrum should be:\n",
    "\n",
    "![img1](MidTerm/kmer1.png)\n",
    "\n",
    "\n",
    "Calling\n",
    "\n",
    "```python\n",
    "myFile = \"test_reads_10k.fasta\"\n",
    "kmers = parseFasta(myFile,17)\n",
    "plotKmerSpectrum(kmers)\n",
    "```\n",
    "\n",
    "should produce:\n",
    "\n",
    "```\n",
    "Read 10000 sequences.\n",
    "Total number of 17mers: 765263\n",
    "```\n",
    "\n",
    "![img2](MidTerm/kmer2_10k.png)\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide a possible solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "\"\"\"Solution to exercise 1\"\"\"\n",
    "\n",
    "def findPositions(S, S1):\n",
    "    s1len = len(S1)\n",
    "    S = S.replace(\"\\n\",\"\")\n",
    "    ret = [x for x in range(0,len(S)) if S[x:x+s1len] == S1]\n",
    "    if len(ret) > 0:\n",
    "        print(\"\\nString '{}' found at the following positions:\".format(S1))\n",
    "    else:\n",
    "        print(\"\\nWarning: string '{}' not found. Returning an empty list.\".format(S1))\n",
    "    return ret\n",
    "\n",
    "\n",
    "S = \"\"\"TGAATACATATTATTCCGC\n",
    "GTCCTACCAAAAAATATACATAGATAT\n",
    "GATACTAGGGGACCGGTTCCGGGATGA\n",
    "TGATGATACTAGGGGACCGGTTGATAC\"\"\"\n",
    "\n",
    "print(findPositions(S, \"TGA\"))\n",
    "print(findPositions(S, \"GAT\"))\n",
    "print(findPositions(S, \"none\"))\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "\"\"\"Solution to exercise 2\"\"\"\n",
    "\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def computeKmers(seq, kmerLen, kmerDict):\n",
    "    \"\"\"\n",
    "    given a DNA sequence seq and an integer kmers size kmerLen,\n",
    "    fills the dictionary kmerDict with all kmers and their multiplicity in the string\n",
    "    \"\"\"\n",
    "    \n",
    "    if(kmerLen < len(seq)):\n",
    "        for x in [seq[i:i + kmerLen] for i in range(len(seq) - kmerLen + 1)]:\n",
    "            if(x in kmerDict):\n",
    "                kmerDict[x] += 1\n",
    "            else:\n",
    "                kmerDict[x] = 1\n",
    "\n",
    "def parseFasta(fileName,kmerLen):\n",
    "    \"\"\"\n",
    "    reads the fasta file fileName\n",
    "    constructs the dictionary with all the kmers\n",
    "    having length kmerLen\n",
    "    and returns the dictionary \n",
    "    \"\"\"\n",
    "\n",
    "    cnt = 0\n",
    "    kmerDict = dict()\n",
    "    for seq_record in SeqIO.parse(fileName, \"fasta\"):\n",
    "        cnt += 1\n",
    "        seq = str(seq_record.seq)\n",
    "        computeKmers(seq, kmerLen, kmerDict)\n",
    "        \n",
    "    print(\"Read {} sequences.\\nTotal number of {}mers: {}\".format(cnt,kmerLen,len(kmerDict)))\n",
    "    \n",
    "    return kmerDict\n",
    "\n",
    "def plotKmerSpectrum(kmerDict):\n",
    "    \"\"\"\n",
    "     Given a dictionary kmerDict (i.e. keys are kmer sequences and \n",
    "     values are the number of occurrences of the kmer sequence) plots \n",
    "     the kmer spectrum of all the kmers present in the dictionary. \n",
    "     The kmer spectrum is a histogram of the multiplicities of all kmers. \n",
    "    \"\"\"\n",
    "\n",
    "    kmerSpect = dict()\n",
    "    for k in kmerDict:\n",
    "        mult = int(kmerDict[k])\n",
    "        oldVal = kmerSpect.get(mult,0) # get multiplicity, if not found returns 0\n",
    "        kmerSpect[mult] = oldVal + 1\n",
    "    \n",
    "    #print(kmerSpect)\n",
    "    listMult = list(kmerSpect.keys())\n",
    "    listMult.sort()\n",
    "    \n",
    "    xVals = []\n",
    "    yVals = []\n",
    "    for val in listMult:\n",
    "            xVals.append(val)\n",
    "            yVals.append(kmerSpect[val])\n",
    "\n",
    "    plt.plot(xVals,yVals)\n",
    "    plt.xlabel(\"Multiplicity\")\n",
    "    plt.ylabel(\"Occurrences\")\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "seq = \"AATTAATTAACTAGCCTTAA\"\n",
    "twoMers = dict()\n",
    "print(\"sequence:\" , seq)\n",
    "computeKmers(seq, 2, twoMers)\n",
    "print(\"twoMers\")\n",
    "print(twoMers)\n",
    "fourMers = dict()\n",
    "computeKmers(seq, 4, fourMers)\n",
    "print(\"4mers\")\n",
    "print(fourMers)\n",
    "t2Mers = dict()\n",
    "computeKmers(seq, 32, t2Mers)\n",
    "print(\"32mers\")\n",
    "print(t2Mers)\n",
    "\n",
    "myFile = \"shortSample.fasta\"\n",
    "kmerDict = parseFasta(myFile,4)\n",
    "print(kmerDict)\n",
    "\n",
    "myFile = \"test_reads_10k.fasta\"\n",
    "kmerDict = parseFasta(myFile,17)\n",
    "plotKmerSpectrum(kmerDict)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "</div>"
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}