{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Module 1, Practical 7\n",
    "\n",
    "In this practical we will see how to get input from the command line. \n",
    "\n",
    "## Gentle reminder on functions\n",
    "\n",
    "Reminder. The basic definition of a function is:\n",
    "```\n",
    "def function_name(input) :\n",
    "    #code implementing the function\n",
    "    ...\n",
    "    ...\n",
    "    return return_value\n",
    "```\n",
    "\n",
    "Functions are defined with the **def** keyword that proceeds the *function_name* and then a list of parameters is passed in the brackets. A colon **:** is used to end the line holding the definition of the function. The code implementing the function is specified by using indentation. A function **might** or **might not** return a value. In the first case a **return** statement is used.\n",
    "\n",
    "\n",
    "## Getting input from the command line\n",
    "\n",
    "To call a program ```my_python_program.py``` from command line, you just have to open a terminal (in Linux) or the command prompt (in Windows) and, assuming that python is present in the path, you can ```cd``` into the folder containing your python program, (eg. ```cd C:\\python\\my_exercises\\```) and just type in \n",
    "```python3 my_python_program.py```\n",
    "or\n",
    "```python my_python_program.py```\n",
    "In case of arguments to be passed by command line, one has to put them after the specification of the program name (eg. ```python my_python_program.py parm1 param2 param3```).\n",
    "\n",
    "Python provides the module **sys** to interact with the interpreter. In particular, **sys.argv** is a list representing all the arguments passed to the python script from the command line.\n",
    "\n",
    "Consider the following code:"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import sys\n",
    "\"\"\"Test input from command line in systest.py\"\"\"\n",
    "\n",
    "if len(sys.argv) != 4: #note that this is the number of params +1!!!\n",
    "    print(\"Dear user, I was expecting 3 params. You gave me \",len(sys.argv)-1)\n",
    "    exit(1)\n",
    "else:\n",
    "    for i in range(0,len(sys.argv)):\n",
    "        print(\"Param {}:{} ({})\".format(i,sys.argv[i],type(sys.argv[i])))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param 2nd 3``` will return:\n",
    "```\n",
    "Param 0: exercises/systest.py (<class 'str'>)\n",
    "Param 1: 1st_param (<class 'str'>)\n",
    "Param 2: 2nd (<class 'str'>)\n",
    "Param 3: 3 (<class 'str'>)\n",
    "```\n",
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param``` will return:\n",
    "```\n",
    "Dear user, I was expecting three parameters. You gave me  1\n",
    "```\n",
    "\n",
    "Note that the parameter at index 0, ```sys.argv[0]``` holds the name of the script, and that all parameters are actually **strings** (and therefore need to be cast to numbers if we want to do mathematical operations on them).\n",
    "\n",
    "**Example:** Write a script that takes two integers in input, i1 and i2, and computes the sum, difference, multiplication and division on them. "
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import sys\n",
    "\"\"\"Maths example with input from command line\"\"\"\n",
    "\n",
    "if len(sys.argv) != 3:\n",
    "    print(\"Dear user, I was expecting 2 params. You gave me \",len(sys.argv)-1)\n",
    "    exit(1)\n",
    "else:\n",
    "    i1 = int(sys.argv[1])\n",
    "    i2 = int(sys.argv[2])\n",
    "    print(\"{} + {} = {}\".format(i1,i2, i1 + i2))\n",
    "    print(\"{} - {} = {}\".format(i1,i2, i1 - i2))\n",
    "    print(\"{} * {} = {}\".format(i1,i2, i1 * i2))\n",
    "    if i2 != 0:\n",
    "        print(\"{} / {} = {}\".format(i1,i2, i1 / i2))\n",
    "    else:\n",
    "        print(\"{} / {} = Infinite\".format(i1,i2))\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Which, depending on user input, should give something like:\n",
    "\n",
    "![i1](img/pract7/math.png)\n",
    "\n",
    "note that we need to check if the values given in input are actually numbers, otherwise the execution will crash (third example). This is easy in case of integers (```str.isdigit()```) but in case of floats it is more complex and might require Exception handling."
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "A more flexible and powerful way of getting input from command line makes use of the ```Argparse``` [module](https://docs.python.org/3/howto/argparse.html). \n",
    "\n",
    "## Argparse\n",
    "\n",
    "Argparse is a command line parsing module which deals with user specified parameters (positional arguments) and optional arguments.\n",
    "\n",
    "\n",
    "Very briefly, the basic syntax of the ```Argparse module``` (for more information check the [official documentation](https://docs.python.org/3/howto/argparse.html)) is the following.\n",
    "\n",
    "1. Import the module:\n",
    "\n",
    "```\n",
    "import argparse\n",
    "```\n",
    "\n",
    "2. Define a argparse object:\n",
    "\n",
    "```\n",
    "parser = argparse.ArgumentParser(description=\"This is the description of the program\")\n",
    "```\n",
    "\n",
    "note the parameter *description* that is a string to describe the program;\n",
    "\n",
    "3. Add positional arguments:\n",
    "```\n",
    "parser.add_argument(\"arg_name\", type = obj, \n",
    "                    help = \"Description of the parameter)\n",
    "```\n",
    "where ```arg_name``` is the name of the argument (which will be used to retrieve its value). The argument has type ```obj``` (the type will be automatically checked for us) and a description specified in the ```help```string.\n",
    "\n",
    "4. Add optional arguments:\n",
    "```\n",
    "parser.add_argument(\"-p\", \"--optional_arg\", type = obj, default = def_val, \n",
    "                        help = \"Description of the parameter)\n",
    "```\n",
    "where ```-p``` is a short form of the parameter (and it is optional), ```--optional_arg``` is the extended name and it requires a value after it is specified, ```type``` is the data type of the parameter passed (e.g. str, int, float, ..), ```default``` is optional and gives a default value to the parameter. If not specified and no argument is passed, the argument will get the value \"None\". ```Help``` is again the description string.\n",
    "\n",
    "5. Parse the arguments:\n",
    "```\n",
    "args = parser.parse_args()\n",
    "```\n",
    "the parser checks the arguments and stores their values in the ```argparse``` object that we called ```args```.\n",
    "\n",
    "6. Retrieve and process arguments:\n",
    "```\n",
    "myArgName = args.arg_name\n",
    "myOptArg = args.optional_arg\n",
    "```\n",
    "now variables contain the values specified by the user and we can use them.\n",
    "\n",
    "Let's see the example above again.\n",
    "\n",
    "**Example:** Write a script that takes two integers in input, i1 and i2, and computes the sum, difference, multiplication and division on them. "
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import argparse\n",
    "\"\"\"Maths example with input from command line\"\"\"\n",
    "parser = argparse.ArgumentParser(description=\"\"\"This script gets two integers  in input \n",
    "and performs some operations on them\"\"\")\n",
    "parser.add_argument(\"i1\", type=int,\n",
    "                    help=\"The first integer\")\n",
    "parser.add_argument(\"i2\", type=int,\n",
    "                    help=\"The second integer\")\n",
    "\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "i1 = args.i1\n",
    "i2 = args.i2\n",
    "print(\"{} + {} = {}\".format(i1,i2, i1 + i2))\n",
    "print(\"{} - {} = {}\".format(i1,i2, i1 - i2))\n",
    "print(\"{} * {} = {}\".format(i1,i2, i1 * i2))\n",
    "if i2 != 0:\n",
    "    print(\"{} / {} = {}\".format(i1,i2, i1 / i2))\n",
    "else:\n",
    "    print(\"{} / {} = Infinite\".format(i1,i2))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "That returns the following:\n",
    "\n",
    "![iap2](img/pract7/systest_argparse.png)\n",
    "\n",
    "\n",
    "Note that we did not have to check the types of the inputs (i.e. the last time we run the script) but this is automatically done by argparse."
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Example:**\n",
    "Let's write a program that gets a string (S) and an integer (N) in input and prints the string repeated N times. Three optional parameters are specified: verbosity (-v) to make the software print a more descriptive output, separator (-s) to separate each copy of the string (defaults to \" \") and trailpoints (-p) to add several \".\" at the end of the string (defaults to 1). "
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import argparse\n",
    "parser = argparse.ArgumentParser(description=\"\"\"This script gets a string \n",
    "                                 and an integer and repeats the string N times\"\"\")\n",
    "parser.add_argument(\"string\", type=str,\n",
    "                    help=\"The string to be repeated\")\n",
    "parser.add_argument(\"N\", type=int,\n",
    "                    help=\"The number of times to repeat the string\")\n",
    "\n",
    "parser.add_argument(\"-v\", \"--verbose\", action=\"store_true\",\n",
    "                    help=\"increase output verbosity\")\n",
    "\n",
    "parser.add_argument(\"-p\", \"--trailpoints\", type = int, default = 1, \n",
    "                    help=\"Adds these many trailing points\")\n",
    "parser.add_argument(\"-s\", \"--separator\", type = str, default = \" \", \n",
    "                    help=\"The separator between repeated strings\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "mySTR = args.string + args.separator\n",
    "trailP = \".\" * args.trailpoints\n",
    "answer = mySTR * args.N \n",
    "\n",
    "answer = answer[:-len(args.separator)] + trailP #to remove the last separator\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"the string {} repeated {} is:\".format(args.str, args.N, answer))\n",
    "else:\n",
    "    print(answer)\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Executing the program from command line without parameters gives the message:\n",
    "\n",
    "![i2](img/pract7/noargs.png)\n",
    "\n",
    "Calling it with the ```-h``` flag:\n",
    "\n",
    "![i3](img/pract7/help.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```:\n",
    "\n",
    "![i4](img/pract7/pos_args.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```, and with the optional parameters ```-s \"___\" -p 3 -v```\n",
    "\n",
    "![i5](img/pract7/sample.png)\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Example:**\n",
    "Let's write a program that reads and prints to screen a text file specified by the user. Optionally, the file might be compressed with gzip to save space. The user should be able to read also gzipped files. Hint: use the module gzip which is very similar to the standard file management method ([more info here](https://docs.python.org/3/library/gzip.html?highlight=gzip#module-gzip)). You can find a text file here [textFile.txt](file_samples/textFile.txt) and its gzipped version here [text.gz](file_samples/textFile.gz):\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import argparse\n",
    "import gzip\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"Reads and prints a text file\"\"\")\n",
    "parser.add_argument(\"filename\", type=str, help=\"The file name\")\n",
    "parser.add_argument(\"-z\", \"--gzipped\", action=\"store_true\", \n",
    "                    help=\"If set, input file is assumed gzipped\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "inputFile = args.filename\n",
    "fh = \"\"\n",
    "if args.gzipped:\n",
    "    fh = gzip.open(inputFile, \"rt\")\n",
    "else:\n",
    "    fh = open(inputFile, \"r\")\n",
    "\n",
    "for line in fh:\n",
    "    line = line.strip(\"\\n\")\n",
    "    print(line)\n",
    "\n",
    "fh.close()\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "The output:\n",
    "\n",
    "![i6](img/pract7/read_gz.png)"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Example:** Let's write a program that reads the content of a file and prints to screen some stats like the number of lines, the number of characters and maximum number of characters in one line. Optionally (if flag -v is set) it should print the content of the file. You can find a text file here [textFile.txt](file_samples/textFile.txt):\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import argparse\n",
    "\n",
    "\n",
    "def readText(f):\n",
    "    \"\"\"reads the file and returns a list with \n",
    "    each line as separate element\"\"\"\n",
    "    myF = open(f, \"r\")\n",
    "    ret = myF.readlines() #careful with big files!\n",
    "    return ret\n",
    "\n",
    "\n",
    "def computeStats(fileList):\n",
    "    \"\"\"returns a tuple (num.lines, num.characters,max_char.line)\"\"\"\n",
    "    num_lines = len(fileList)\n",
    "    lines_len = [len(x.replace(\"\\n\", \"\")) for x in fileList]\n",
    "    num_char = sum(lines_len)\n",
    "    max_char = max(lines_len)\n",
    "    return (num_lines, num_char, max_char)\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Computes file stats\")\n",
    "parser.add_argument(\"inputFile\", type=str, help=\"The input file\")\n",
    "parser.add_argument(\n",
    "    \"-v\", \"--verbose\", action=\"store_true\", help=\"if set, prints the file content\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "inFile = args.inputFile\n",
    "\n",
    "lines = readText(inFile)\n",
    "stats = computeStats(lines)\n",
    "if args.verbose:\n",
    "    print(\"File content:\\n{}\\n\".format(\"\".join(lines)))\n",
    "print(\n",
    "    \"Stats:\\nN.lines:{}\\nN.chars:{}\\nMax. char in line:{}\".format(\n",
    "        stats[0], stats[1], stats[2]))\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Output with -v flag:\n",
    "\n",
    "![i7](img/pract7/fileread.png)\n",
    "\n",
    "Output without -v flag:\n",
    "\n",
    "![i8](img/pract7/filereadnov.png)"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Libraries installation for the next practicals\n",
    "\n",
    "This section is not mandatory for the current practical, but set the basis for the next ones. For perfoming the next practicals we will need three additional libraries. Try and see if they are already available by typing the following commands in the console or put them in a python script:\n",
    "```\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "```\n",
    "\n",
    "if, upon execution, you do not get any error messages, you are sorted. Otherwise, you need to install them.\n",
    "\n",
    "In **Linux** you can install the libraries by typing in a terminal ```sudo pip3 install matplotlib```, ```sudo pip3 install pandas``` and ```sudo pip3 install numpy``` (or ```sudo python3.X -m pip install matplotlib```, ```sudo python3.X -m pip install pandas``` and ```sudo python3.X -m pip install numpy```), where X is your python version.\n",
    "\n",
    "In **Windows** you can install the libraries by typing in the command prompt (to open it type ```cmd``` in the search box) ```pip3 install matplotlib```, ```pip3 install pandas``` and ```pip3 install numpy```. If you are using anaconda you need to run these commands from the *anaconda prompt*.\n",
    "\n",
    "**Please install them in this order** (i.e. **matplotlib** first, then **pandas** and finally **numpy**). You might not need to install numpy as matplotlib requires it. Once done that, try to perform the above imports again and they should work this time around.\n",
    "\n",
    "## Exercises\n",
    "\n",
    "1. Modify the program of Exercise 6 of Practical 6 in order to allow users to specify the input and output files from command line. Then test it with the provided files. The text of the exercise follows:\n",
    "\n",
    "Write a python program that reads two files. The first is a one column text file ([contig_ids.txt](file_samples/contig_ids.txt)) with the identifiers of some contigs that are present in the second file, which is a fasta formatted file ([contigs82.fasta](file_samples/contigs82.fasta)). The program will write on a third, fasta formatted file (e.g. filtered_contigs.fasta) only those entries in *contigs82.fasta* having identifier in *contig_ids.txt*.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "\"\"\"exercises/filterFasta.py\"\"\"\n",
    "\n",
    "import argparse\n",
    "\n",
    "def readIDS(f):\n",
    "    \"\"\"reads a one column file in and stores\n",
    "    the ids in a dictionary that is returned at the end\"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line not in ret):\n",
    "                ret[line] = 1 #Important. It is like: True\n",
    "    return ret\n",
    "\n",
    "def filterFasta(inF, outF, ids2keep):\n",
    "    oF = open(outF, \"w\")\n",
    "    \n",
    "    outputME = False\n",
    "    with open(inF, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                #this is the header\n",
    "                if ids2keep.get(line[1:],False):\n",
    "                    oF.write(line +\"\\n\")\n",
    "                    outputME = True\n",
    "                    print(\"Writing contig \", line[1:])\n",
    "                else:\n",
    "                    outputME = False\n",
    "            else:\n",
    "                if outputME:\n",
    "                    oF.write(line +\"\\n\")\n",
    "        \n",
    "    oF.close()\n",
    "    \n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Filters a fasta file\")\n",
    "parser.add_argument(\"inputFasta\", type = str, help = \"The input fasta file\")\n",
    "parser.add_argument(\"inputIDS\", type = str, help = \"The IDS to keep\")\n",
    "parser.add_argument(\"outputFasta\", type = str, \n",
    "                    help = \"The output fasta file with filtered entries\")\n",
    "args = parser.parse_args()\n",
    "idsFile = args.inputIDS\n",
    "inFasta = args.inputFasta\n",
    "outFasta = args.outputFasta\n",
    "\n",
    "ids = readIDS(idsFile)\n",
    "filterFasta(inFasta,outFasta, ids)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "</div>"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "2. Write a python script that takes in input a single-entry .fasta file (specified from the command line) of the amino-acidic sequence of a protein and prints off (1) the total number of aminoacids, (2) for each aminoacid, its count and percentage of the whole. Optionally, it the user specifies the flag \"-S\" (--search) followed by a string representing an aminoacid sequence, the program should count and print how many times that input sequence appears. Download the [Sars-Cov-2 Spike Protein](file_samples/P0DTC2.fasta.txt) and test your script on it. *Please use functions*. \n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1-cov');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1-cov\" style=\"display:none;\">"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "\"\"\"exercises/process_fasta.py\"\"\"\n",
    "\n",
    "\"\"\" test example: \n",
    "python3 process_fasta.py ../file_samples/P0DTC2.fasta.txt -S SSVL \"\"\"\n",
    "\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Parses a single-entry fasta file and returns some stats.\")\n",
    "parser.add_argument(\"inputFasta\", type = str, help = \"The input fasta file\")\n",
    "parser.add_argument(\"-S\", \"--search\", type = str, default = \"\", \n",
    "                    help=\"The (optional) string to look for.\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Reads a fasta file input_file in \n",
    "#and returns the tuple (header, sequence)\n",
    "def read_sequence(input_file):\n",
    "    sequence = \"\"\n",
    "    hdr = \"\"\n",
    "    inF = open(input_file)\n",
    "    for line in inF:\n",
    "        line = line.strip()\n",
    "        if not line.startswith(\">\"):\n",
    "            sequence += line\n",
    "        else:\n",
    "            hdr = line[1:]\n",
    "\n",
    "    return hdr,sequence\n",
    "\n",
    "\n",
    "# Gets a sequence seq and returns \n",
    "# a dictionary with the counts of all elements.\n",
    "# This function also prints off the counts (and %)\n",
    "def count_chars(seq):\n",
    "    char_dict = dict()\n",
    "    L = len(seq)\n",
    "    for c in seq:\n",
    "        if char_dict.get(c, None) == None:\n",
    "            char_dict[c] = 0\n",
    "        char_dict[c] += 1\n",
    "    \n",
    "    print(\"The sequence has length: {}\".format(L)) \n",
    "    for el in char_dict:\n",
    "        print(\"{} is present {} times ({:.2f} %)\".format(el, char_dict[el], 100 * char_dict[el]/L))\n",
    "    \n",
    "    return char_dict\n",
    "\n",
    "# Counts how many times search_s is in seq and returns an integer\n",
    "def count_str(seq, search_s):\n",
    "    return seq.count(search_s)\n",
    "   \n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "inFasta = args.inputFasta\n",
    "\n",
    "search_str = args.search\n",
    "\n",
    "\n",
    "h,s = read_sequence(inFasta)\n",
    "print(h)\n",
    "print(s)\n",
    "cnts = count_chars(s)\n",
    "\n",
    "if len(search_str) > 0:\n",
    "    C = count_str(s, search_str)\n",
    "    print(\"Sequence '{}' is present {} times \".format(search_str, C))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "The output should be like:\n",
    "```    \n",
    "The sequence has length: 1273\n",
    "M is present 14 times (1.10 %)\n",
    "F is present 77 times (6.05 %)\n",
    "V is present 97 times (7.62 %)\n",
    "L is present 108 times (8.48 %)\n",
    "P is present 58 times (4.56 %)\n",
    "S is present 99 times (7.78 %)\n",
    "Q is present 62 times (4.87 %)\n",
    "C is present 40 times (3.14 %)\n",
    "N is present 88 times (6.91 %)\n",
    "T is present 97 times (7.62 %)\n",
    "R is present 42 times (3.30 %)\n",
    "A is present 79 times (6.21 %)\n",
    "Y is present 54 times (4.24 %)\n",
    "G is present 82 times (6.44 %)\n",
    "D is present 62 times (4.87 %)\n",
    "K is present 61 times (4.79 %)\n",
    "H is present 17 times (1.34 %)\n",
    "W is present 12 times (0.94 %)\n",
    "I is present 76 times (5.97 %)\n",
    "E is present 48 times (3.77 %)\n",
    "Sequence 'SSVL' is present 2 times\n",
    "```"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "</div>"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "3. Given a [fasta](https://en.wikipedia.org/wiki/FASTA_format) file like [contigs82.fasta](file_samples/contigs82.fasta) specified in input by a user,  write a python script that counts, for each sequence, the number of times that a DNA or protein string specified in input appears. \n",
    "\n",
    "If we run something like: python3 ```find_stringInFasta.py contigs82.fasta TGCTCACAG```\n",
    "\n",
    "the result should print lines like:\n",
    "```\n",
    "TGCTCACAG in MDC052568.000: 1 times\n",
    "TGCTCACAG in MDC002479.192: 1 times\n",
    "TGCTCACAG in MDC040033.7: 1 times\n",
    "```\n",
    "\n",
    "Modify the program so that it outputs also the list of all the indexes where the string appears in each sequence in the fasta file. Try to look for the following sequences:\n",
    "```\n",
    "TTTTCCTAGG\n",
    "TGCTCCGAGCATGTGATAATCATTCCAAGCTCCAT\n",
    "TAAACAT\n",
    "GATTACA\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "\"\"\"exercises/find_stringInFasta.py\"\"\"\n",
    "\n",
    "\n",
    "import argparse\n",
    "\n",
    "def getIndexes(string1,string2):\n",
    "    \"\"\"checks if string2 is present in string1 and returns \n",
    "    all the positions at which string2 occurs in string1\"\"\"\n",
    "    ret = []\n",
    "    ind = string1.find(string2)\n",
    "    \n",
    "    while (ind > -1 and ind < len(string1)):\n",
    "        ret.append(ind)\n",
    "        ind = string1.find(string2,ind + 1)\n",
    "        \n",
    "    return ret\n",
    "\n",
    "def processFasta(file,testStr):\n",
    "    \"\"\"reads a fasta file entry by entry checks if the input string\n",
    "       testStr is present in each sequence. Reporting how many times and where.\n",
    "    \"\"\"\n",
    "    header = \"\"\n",
    "    seq = \"\"\n",
    "    with open(file, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                if(len(header) == 0 ):\n",
    "                    #first entry:\n",
    "                    header = line[1:]\n",
    "                else:\n",
    "                    #this is a new entry\n",
    "                    indexes = getIndexes(seq,testStr)\n",
    "                    if len(indexes) > 0:\n",
    "                        print(\"{} in {}: {} times ({})\".format(testStr, \n",
    "                                                               header,\n",
    "                                                               len(indexes),\n",
    "                                                               indexes))\n",
    "                    seq = \"\"\n",
    "                    header = line[1:]\n",
    "            else:\n",
    "                seq +=line\n",
    "    #processing the final entry\n",
    "    indexes = getIndexes(seq,testStr)\n",
    "    if len(indexes) > 0:\n",
    "        print(\"{} in {}: {} times ({})\".format(testStr, \n",
    "                                               header,\n",
    "                                               len(indexes),\n",
    "                                               indexes))\n",
    "                    \n",
    "parser = argparse.ArgumentParser(\n",
    "    description=\"\"\"Checks if a sequence is exactly contained in a fasta file\"\"\"\n",
    ")\n",
    "parser.add_argument(\"filename\", type=str, help=\"The fasta file name\")\n",
    "parser.add_argument(\"query\", type=str, help=\"The query string\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "#inFasta = args.filename\n",
    "#testS = args.query\n",
    "inFasta = \"file_samples/contigs82.fasta\"\n",
    "testS = \"TAAACAT\"\n",
    "processFasta(inFasta, testS)"
   ],
   "outputs": [],
   "metadata": {
    "scrolled": true
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "</div>"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "4. The [Fisher's dataset](http://onlinelibrary.wiley.com/doi/10.1111/j.1469-1809.1936.tb02137.x/abstract) regarding Petal and Sepal length and width in csv format can be found [here](file_samples/Fishers_Iris.csv). These are the measurements of  the flowers of  fifty plants each of the two species Iris setosa and Iris versicolor.\n",
    "\n",
    "The header of the file is:\n",
    "```\n",
    "Species Number,Species Name,Petal width,Petal length,Sepal length,Sepal width\n",
    "```\n",
    "\n",
    "Write a python script that reads this file in input (feel free to hard-code the filename in the code) and computes the average petal length and width and sepal length and width for each of the three different Iris species. Print them to the screen alongside the number of elements.\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "def readCSV(f):\n",
    "    \"\"\"reads the csv dataset and returns a dictionary with \n",
    "    species name as key and, as value, a dictionary with \n",
    "    four keys : petalLen, sepalLen, petalWidth, sepalWidth\n",
    "    \"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if not line.startswith(\"Species Number\"):\n",
    "                data = line.split(\",\")\n",
    "                speciesName = data[1]\n",
    "                pWidth = int(data[2])\n",
    "                pLen = int(data[3])\n",
    "                sLen = int(data[4])\n",
    "                sWidth = int(data[5])\n",
    "                if(speciesName not in ret):\n",
    "                    ret[speciesName] = {\"petalLen\" : [], \"sepalLen\" : [], \n",
    "                                        \"petalWidth\" : [], \"sepalWidth\" : []\n",
    "                                       }\n",
    "                ret[speciesName][\"petalLen\"].append(pLen)\n",
    "                ret[speciesName][\"sepalLen\"].append(sLen)\n",
    "                ret[speciesName][\"sepalWidth\"].append(sWidth)\n",
    "                ret[speciesName][\"petalWidth\"].append(pWidth)\n",
    "    return ret\n",
    "\n",
    "def printData(dataDict):\n",
    "    for s in dataDict:\n",
    "        avgPlen = sum(dataDict[s][\"petalLen\"])/len(dataDict[s][\"petalLen\"])\n",
    "        avgPwid = sum(dataDict[s][\"petalWidth\"])/len(dataDict[s][\"petalWidth\"])\n",
    "        avgSlen = sum(dataDict[s][\"sepalLen\"])/len(dataDict[s][\"sepalLen\"])\n",
    "        avgSwid = sum(dataDict[s][\"sepalWidth\"])/len(dataDict[s][\"sepalWidth\"])\n",
    "        print(\"Species {} has {} measurements:\".format(s, len(dataDict[s][\"petalLen\"])))\n",
    "        print(\"\\t petal length {}\".format(avgPlen))\n",
    "        print(\"\\t petal width {}\".format(avgPwid))\n",
    "        print(\"\\t sepal length {}\".format(avgSlen))\n",
    "        print(\"\\t sepal width {}\".format(avgSwid))\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "inFile = \"file_samples/Fishers_Iris.csv\"\n",
    "\n",
    "data = readCSV(inFile)\n",
    "printData(data)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "</div>"
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}